<h1><center>Лабораторна робота 10.</center></h1>
<h2><center>
    Прогнозування затримок вильоту за допомогою градієнтного підсилення
</center></h2>

**Виконав:** Прізвище І.П.

**Варіант:** №__

Ваше завдання – перевершити щонайменше два бенчмарки в цьому [навчальному змаганні на Kaggle](https://www.kaggle.com/c/flight-delays-spring-2018/leaderboard).

Нижче наведено короткий опис того, як було досягнуто другого бенчмарку за допомогою XGBoost. На цьому етапі курсу «Інтелектуальний аналіз даних» вам достатньо поверхнево ознайомитися з даними, щоб зрозуміти: це саме той тип задачі, у якому градієнтне підсилення працює особливо добре. 

Швидше за все, ви теж використаєте XGBoost, однак майте на увазі, що в наданому наборі даних є чимало категоріальних ознак. Тому дотримуйтеся наведених нижче інструкцій та рекомендацій, щоб досягти кращого результату.


<figure>
  <img src="https://raw.githubusercontent.com/radiukpavlo/intelligent-data-analysis/refs/heads/main/03_img/10_5_xgboost-2.jpeg" align="center" width="50%" alt="Візуалізація моделі XGBoost">
  <figcaption> 
    <a href="https://community.ultralytics.com/t/a-new-meme-2024-paris-olympics/160">Джерело</a>
  </figcaption>
</figure>


<a class="anchor" id="lab-10"></a>

## Зміст

- [10.1. Дослідження даних та попередня обробка з градієнтним підсиленням](#lab-10.1)
- [10.2. Конструювання ознак та зменшення розмірності](#lab-10.2)
- [10.3. Налаштування гіперпараметрів та оцінювання моделі](#lab-10.3)
- [10.4. Обробка пропущених даних, викидів та особливих випадків](#lab-10.4)
- [10.5. Розширені ансамблеві стратегії та інтерпретація](#lab-10.5)


In [1]:
import warnings

warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from xgboost import XGBClassifier

In [2]:
train = pd.read_csv('https://raw.githubusercontent.com/radiukpavlo/intelligent-data-analysis/refs/heads/main/02_assignments/ida_lab-10_flight-delays-kaggle/flight_delays_train.csv')
test = pd.read_csv('https://raw.githubusercontent.com/radiukpavlo/intelligent-data-analysis/refs/heads/main/02_assignments/ida_lab-10_flight-delays-kaggle/flight_delays_test.csv')

In [3]:
train.head()

,Month,DayofMonth,DayOfWeek,DepTime,UniqueCarrier,Origin,Dest,Distance,dep_delayed_15min
0,c-8,c-21,c-7,1934,AA,ATL,DFW,732,N
1,c-4,c-20,c-3,1548,US,PIT,MCO,834,N
2,c-9,c-2,c-5,1422,XE,RDU,CLE,416,N
3,c-11,c-25,c-6,1015,OO,DEN,MEM,872,N
4,c-10,c-7,c-6,1828,WN,MDW,OMA,423,Y


In [4]:
test.head()

,Month,DayofMonth,DayOfWeek,DepTime,UniqueCarrier,Origin,Dest,Distance
0,c-7,c-25,c-3,615,YV,MRY,PHX,598
1,c-4,c-17,c-2,739,WN,LAS,HOU,1235
2,c-12,c-2,c-7,651,MQ,GSP,ORD,577
3,c-3,c-25,c-7,1614,WN,BWI,MHT,377
4,c-6,c-6,c-3,1505,UA,ORD,STL,258


Маючи час вильоту, код авіаперевізника, аеропорт відправлення, пункт призначення та дальність польоту, потрібно спрогнозувати, чи перевищить затримка вильоту 15 хвилин. За простий бенчмарк візьмемо класифікатор XGBoost і дві ознаки, які використати найпростіше: `DepTime` та `Distance`. Така модель дає результат $0.68202$ на таблиці лідерів.


In [5]:
X_train = train[['Distance', 'DepTime']].values
y_train = train['dep_delayed_15min'].map({'Y': 1, 'N': 0}).values
X_test = test[['Distance', 'DepTime']].values

X_train_part, X_valid, y_train_part, y_valid = train_test_split(
    X_train, y_train, test_size=0.3, random_state=17
)

Навчимо XGBoost з параметрами за замовчуванням на частині даних і оцінимо значення `ROC AUC` на відкладеній вибірці.


In [6]:
xgb_model = XGBClassifier(seed=17)

xgb_model.fit(X_train_part, y_train_part)
xgb_valid_pred = xgb_model.predict_proba(X_valid)[:, 1]

roc_auc_score(y_valid, xgb_valid_pred)

0.7001348346148775

Тепер виконаємо те саме на всьому тренувальному наборі, побудуємо прогнози для тестового набору й сформуємо файл подання. Саме так можна перевершити перший бенчмарк.


In [7]:
xgb_model.fit(X_train, y_train)
xgb_test_pred = xgb_model.predict_proba(X_test)[:, 1]

pd.Series(xgb_test_pred, name='dep_delayed_15min').to_csv(
    'xgb_2feat.csv', index_label='id', header=True
)

Другий бенчмарк на таблиці лідерів було досягнуто так:

- Ознаки `Distance` і `DepTime` були використані без змін.
- Ознаку `Flight` було створено з ознак `Origin` і `Dest`.
- Ознаки `Month`, `DayofMonth`, `DayOfWeek`, `UniqueCarrier` і `Flight` було перетворено за допомогою OHE (`LabelBinarizer`).
- Було навчено логістичну регресію та градієнтне підсилення (XGBoost). Гіперпараметри XGBoost налаштовувалися за допомогою крос-валідації. Спочатку оптимізували параметри, що відповідають за складність моделі, а потім фіксували кількість дерев на рівні 500 і налаштовували крок навчання.
- Імовірності прогнозувалися за допомогою крос-валідації з використанням `cross_val_predict`. Лінійну суміш прогнозів логістичної регресії та градієнтного підсилення було задано у вигляді $w_1 * p_{logit} + (1 - w_1) * p_{xgb}$, де $p_{logit}$ – імовірність класу 1, передбачена логістичною регресією, а $p_{xgb}$ – відповідна імовірність, передбачена XGBoost. Вагу $w_1$ добирали вручну.
- Аналогічну комбінацію прогнозів було сформовано і для тестового набору.

Дотримуватися саме цих кроків не обов’язково. Це лише опис того, як автор цього завдання отримав результат. Можливо, ви не захочете повторювати той самий підхід і, наприклад, додасте кілька вдалих ознак та навчите випадковий ліс із тисячі дерев. У будь-якому разі виконайте наведені нижче завдання, щоб досягти найкращого результату.


<a class="anchor" id="lab-10.1"></a>

## <span style="color:blue; font-size:1.5em;">10.1. Дослідження даних та попередня обробка з градієнтним підсиленням</span>

[Назад до змісту](#lab-10)


### <span style='color:red; font-size:1.4em;'>Завдання 1</span>

**Для всіх варіантів:**

- Працюйте лише з даними `flight_delays_train.csv` і `flight_delays_test.csv`; цільова змінна `dep_delayed_15min` наявна лише у `train`, тому якість моделі оцінюйте на валідаційній вибірці або через крос-валідацію на `train`.
- Перед моделюванням приведіть `Month`, `DayofMonth` і `DayOfWeek` до числового вигляду, коректно опрацювавши службовий префікс `c-`, а також перевірте допустимість значень `DepTime` і `Distance`.
- Усі правила очищення, які не використовують `dep_delayed_15min`, застосовуйте однаково до `train` і `test`; у звіті обов’язково вкажіть кількість виправлених, вилучених або заповнених записів.
- Для кожного варіанта подайте щонайменше одну таблицю порівняння, одну візуалізацію або матрицю помилок і стислий висновок про вплив підготовки даних на `roc_auc_score`, `f1_score` або `accuracy_score`.

---

**Варіант 1 – Первинне очищення і базова модель на `DepTime` та `Distance`:**

- **Мета:** побудувати першу відтворювану baseline-модель після мінімально необхідного очищення даних.

- **Кроки:**
    1. Завантажте `flight_delays_train.csv` і `flight_delays_test.csv`, перевірте типи стовпців та перетворіть `Month`, `DayofMonth`, `DayOfWeek` зі строкового формату `c-*` у цілі числа.
    2. Переконайтеся, що `DepTime` лежить у межах від `0` до `2359`, а `Distance` є додатною; окремо порахуйте записи з некоректними значеннями та оберіть правило їх обробки.
    3. Сформуйте baseline-ознаки лише з `DepTime` і `Distance`, перетворіть `dep_delayed_15min` з `Y/N` у `1/0` та виконайте `train_test_split` з фіксованим `random_state`.
    4. Навчіть `XGBoost` або `XGBClassifier` з базовими параметрами, обчисліть `accuracy_score`, `roc_auc_score` і побудуйте confusion matrix для валідаційної частини.
    5. Порівняйте якість до та після очищення, поясніть, які саме помилки формату найбільше вплинули на стійкість baseline-моделі.

- **Підказки:** використайте `str.replace('c-', '', regex=False)` для календарних полів, а ціль перекодуйте через `map({'Y': 1, 'N': 0})`.

---

**Варіант 2 – Календарні ознаки та базове кодування дати:**

- **Мета:** з’ясувати, чи покращують прості календарні ознаки початковий прогноз затримки.

- **Кроки:**
    1. Очистіть `Month`, `DayofMonth` і `DayOfWeek`, перевірте їхні діапазони та сформуйте числову версію календарного блоку без втрати записів.
    2. Створіть простий індикатор часу доби на основі `DepTime`, наприклад `early_morning`, `morning`, `afternoon`, `evening`, і зафіксуйте спосіб кодування.
    3. Побудуйте дві моделі: baseline лише з `DepTime` і `Distance`, а також розширену модель з доданими календарними ознаками та індикатором часу доби.
    4. Оцініть обидві моделі за однаковим розбиттям даних, наведіть `roc_auc_score` або `f1_score` та порівняйте важливість ознак.
    5. Сформулюйте висновок, чи дає календарна структура реальний виграш уже на базовому етапі підготовки.

- **Підказки:** для простого кодування використайте `pandas.get_dummies()` або залиште календарні поля числовими й поясніть вибір.

---

**Варіант 3 – Контроль викидів `Distance` і перевірка масштабування:**

- **Мета:** оцінити, чи змінюють викиди у `Distance` і масштабування числових ознак якість дерева градієнтного підсилення.

- **Кроки:**
    1. Дослідіть розподіл `Distance`, побудуйте гістограму та обчисліть принаймні 95-й і 99-й перцентилі для виявлення підозріло великих значень.
    2. Сформуйте дві версії навчального набору: з первинними даними та з обмеженням або winsorization для екстремальних відстаней.
    3. За потреби масштабування створіть третю версію даних зі `StandardScaler` лише для числових полів `DepTime` і `Distance`.
    4. Навчіть однакові `XGBoost`-моделі на кожній версії набору й порівняйте метрики, а також, за можливості, зміну часу навчання.
    5. Поясніть, чи справді масштабування допомогло, чи основний ефект був пов’язаний саме з контролем викидів.

- **Підказки:** для tree-based моделей масштабування часто не є критичним, тому аргументуйте висновок числовими результатами, а не загальним правилом.

---

**Варіант 4 – Дедуплікація та перевірка стабільності baseline-моделі:**

- **Мета:** встановити, чи впливають дублікати записів на якість і стабільність прогнозування.

- **Кроки:**
    1. Визначте ключ дубліката на основі `Month`, `DayofMonth`, `DayOfWeek`, `DepTime`, `UniqueCarrier`, `Origin`, `Dest` і `Distance`, порахуйте кількість повторів.
    2. Створіть дві версії навчального набору: вихідну і deduplicated, зберігши однакові перетворення для всіх інших ознак.
    3. Для кожної версії навчіть однакову `XGBoost`-модель, використавши `cross_val_score` або ручну 5-fold-крос-валідацію.
    4. Порівняйте середнє значення метрики та стандартне відхилення між двома сценаріями, а також зміну розподілу класів після вилучення дублікатів.
    5. Зробіть висновок, чи дублікати створювали штучну перевагу моделі або, навпаки, не змінювали результат істотно.

- **Підказки:** якщо дублікати мають різні мітки `dep_delayed_15min`, опишіть це окремо як потенційний прояв шуму в даних.

---

**Варіант 5 – Пропуски і службові коди в календарних ознаках:**

- **Мета:** порівняти вилучення та імпутацію записів з пошкодженими календарними даними.

- **Кроки:**
    1. Перевірте `Month`, `DayofMonth`, `DayOfWeek` і `DepTime` на порожні значення, службові коди, нечислові символи або значення поза допустимим діапазоном.
    2. Реалізуйте дві стратегії: вилучення проблемних рядків і корекцію або заповнення значень, наприклад медіаною для `DepTime`.
    3. Побудуйте для кожної стратегії однаковий набір ознак і навчіть `XGBoost` з фіксованими гіперпараметрами.
    4. Порівняйте `f1_score`, `roc_auc_score`, кількість збережених рядків і зміну частки класу `Y`.
    5. Поясніть, чому саме одна зі стратегій є кращою для подальших етапів лабораторної роботи.

- **Підказки:** якщо виправляєте формат календарних полів, окремо зафіксуйте, скільки значень було очищено простим видаленням префікса `c-`.

---

**Варіант 6 – Випадковий і хронологічний поділ вибірки:**

- **Мета:** перевірити, чи не переоцінює випадковий split якість моделі порівняно з наближеним часовим сценарієм.

- **Кроки:**
    1. Після очищення календарних полів відсортуйте `train` за `Month`, `DayofMonth` і `DepTime`, щоб відтворити наближений хронологічний порядок.
    2. Сформуйте два способи поділу: випадковий через `train_test_split` і послідовний, де ранні спостереження становлять train, а пізніші – validation.
    3. Навчіть одну й ту саму `XGBoost`-модель на кожному поділі без зміни гіперпараметрів і набору ознак.
    4. Порівняйте `roc_auc_score`, `accuracy_score` або `f1_score`, а також розподіл місяців і годин вильоту між train та validation у кожному сценарії.
    5. Обґрунтуйте, який спосіб валідації є більш чесним для задачі прогнозування майбутніх затримок.

- **Підказки:** у висновку окремо прокоментуйте ризик leakage, якщо в train і validation потрапляють дуже схожі записи з однаковими календарними шаблонами.

---

**Варіант 7 – Профіль доби та очищення аномалій `DepTime`:**

- **Мета:** пов’язати часову структуру вильотів із початковою якістю baseline-моделі.

- **Кроки:**
    1. Перетворіть `DepTime` на години і хвилини, побудуйте розподіл вильотів за годинами та окремо обчисліть частку затримок у кожній годині.
    2. Виявте аномальні значення часу, наприклад від’ємні коди або значення понад `2359`, і сформуйте правило їх корекції або відсікання.
    3. Навчіть baseline-модель до очищення `DepTime` і повторіть навчання після усунення аномалій.
    4. Порівняйте метрики двох моделей і покажіть, чи змінилася форма часового профілю після очищення.
    5. Поясніть, які часові інтервали найкраще сигналізують про підвищений ризик затримки.

- **Підказки:** для візуалізації підійдуть гістограма, стовпчикова діаграма за годинами та окремий графік частки затримок за годинами.

---

**Варіант 8 – Індикатор пікових календарних періодів без зовнішніх даних:**

- **Мета:** отримати просту календарну ознаку ризику, не залучаючи сторонніх джерел.

- **Кроки:**
    1. На основі `Month`, `DayofMonth` і `DayOfWeek` створіть кілька простих календарних прапорців, наприклад `is_month_start`, `is_month_end`, `is_weekend`.
    2. Дослідіть, чи відрізняється частка затримок для кожного прапорця порівняно з усім набором даних.
    3. Додайте найінформативніші прапорці до baseline-ознак і навчіть розширену `XGBoost`-модель.
    4. Порівняйте результат з моделлю без календарних прапорців, використовуючи однаковий split і ту саму метрику.
    5. Обґрунтуйте, чи є сенс включати такі прості календарні проксі-ознаки вже на стартовому етапі.

- **Підказки:** не вигадуйте зовнішні свята; спирайтеся лише на шаблони, які можна обчислити з наявних колонок.

---

**Варіант 9 – Великі вузлові аеропорти проти інших:**

- **Мета:** перевірити, чи дає розділення на hub і non-hub додаткову аналітичну цінність для baseline-моделі.

- **Кроки:**
    1. За кількістю рейсів у `Origin` визначте п’ять найбільших аеропортів відправлення та позначте всі інші як окрему групу.
    2. Порівняйте між групами розподіл `DepTime`, `Distance` і частку `dep_delayed_15min = Y`.
    3. Навчіть окремі baseline-моделі для hub- та non-hub-підмножин і паралельно побудуйте глобальну модель на всіх даних.
    4. Оцініть якість кожного підходу на відповідних валідаційних частинах і порівняйте, чи виправдане спеціалізоване моделювання.
    5. Зробіть висновок, чи відрізняються закономірності затримок для великих аеропортів настільки, щоб будувати окремі моделі вже на старті.

- **Підказки:** якщо даних у non-hub-групі дуже багато, використайте однакову схему вибірки, щоб порівняння не було викривленим.

---

**Варіант 10 – `Label Encoding` проти `One-Hot Encoding` для baseline-моделі:**

- **Мета:** порівняти два базові способи кодування категоріальних полів у задачі затримок рейсів.

- **Кроки:**
    1. Очистіть числові поля і підготуйте категоріальні стовпці `UniqueCarrier`, `Origin` та `Dest` для двох схем кодування.
    2. Створіть перший набір ознак із `LabelEncoder`, а другий – з `OneHotEncoder` або `pandas.get_dummies()`, зберігши `DepTime` і `Distance`.
    3. Навчіть однакову `XGBoost`-модель на кожному представленні та зафіксуйте час навчання, кількість ознак і споживання пам’яті.
    4. Порівняйте метрики двох моделей на спільній валідації та перевірте, чи не з’являється суттєвий дисбаланс між швидкістю і якістю.
    5. Сформулюйте висновок, який спосіб кодування є кращим компромісом для подальших етапів лабораторної роботи.

- **Підказки:** якщо застосовуєте `LabelEncoder`, обов’язково поясніть ризик штучного порядку між категоріями.

---

**Варіант 11 – Об’єднання рідкісних категорій у `Other`:**

- **Мета:** зменшити фрагментацію категоріальних ознак без втрати основної інформації.

- **Кроки:**
    1. Порахуйте частоти категорій у `UniqueCarrier`, `Origin` і, за потреби, `Dest`, після чого виберіть поріг, нижче якого категорія вважається рідкісною.
    2. Побудуйте нову версію даних, де всі рідкісні значення об’єднані в категорію `Other`, і закодуйте її так само, як вихідний набір.
    3. Навчіть baseline-модель на початковому наборі та окремо на наборі з об’єднаними категоріями.
    4. Порівняйте якість, кількість ознак і зрозумілість отриманих важливостей ознак.
    5. Обґрунтуйте, чи справді зменшення категоріальної розмірності покращує узагальнення моделі.

- **Підказки:** поріг частоти добирайте не довільно, а з огляду на розмір train і частку унікальних значень у кожному стовпці.

---

**Варіант 12 – Парні взаємодії ознак у початковому наближенні:**

- **Мета:** перевірити, чи додають прості interaction features корисний сигнал уже на базовому рівні.

- **Кроки:**
    1. Оберіть дві або три змістовні взаємодії, наприклад `Origin_Dest`, `DepHour_DayOfWeek` або `DistanceBin_Month`, після попереднього очищення даних.
    2. Перетворіть обрані взаємодії на категоріальні або числові ознаки, не змінюючи решту baseline-блоку.
    3. Навчіть модель без interaction features і модель з interaction features за однаковими гіперпараметрами.
    4. Порівняйте метрики, а також ранжування важливостей ознак, щоб побачити, чи потрапили взаємодії до числа найвпливовіших.
    5. Поясніть, які комбінації ознак є найзмістовнішими саме для затримок вильоту.

- **Підказки:** не створюйте надто багато взаємодій одразу; краще порівняти кілька обґрунтованих комбінацій і пояснити їхній вплив.

---

**Варіант 13 – Поліноміальні перетворення числових ознак:**

- **Мета:** з’ясувати, чи корисні квадратичні та перехресні числові ознаки для baseline-моделі з градієнтним підсиленням.

- **Кроки:**
    1. Після очищення `DepTime` перетворіть його на більш інтерпретовану форму, наприклад `DepHour`, і підготуйте `Distance` як другу базову числову ознаку.
    2. Створіть поліноміальні похідні `DepHour^2`, `Distance^2` і `DepHour * Distance` вручну або через `PolynomialFeatures`.
    3. Навчіть baseline-модель без поліноміальних ознак і модель з розширеним числовим блоком.
    4. Порівняйте якість, час навчання та, за наявності, важливості ознак для нових членів полінома.
    5. Зробіть висновок, чи дають поліноміальні перетворення додаткову користь, якщо дерево підсилення і без того моделює нелінійності.

- **Підказки:** тримайте ступінь полінома не вище другого, інакше порівняння швидко втратить інтерпретованість.

---

**Варіант 14 – Ознака великого аеропорту відправлення:**

- **Мета:** оцінити вплив простої бінарної ознаки `isMajorOrigin` на якість початкового прогнозу.

- **Кроки:**
    1. За кількістю спостережень у `Origin` визначте перелік найбільших аеропортів відправлення та сформуйте бінарний прапорець `isMajorOrigin`.
    2. Порівняйте частку затримок між великими аеропортами та рештою, використавши зведену таблицю або стовпчикову діаграму.
    3. Додайте `isMajorOrigin` до baseline-ознак разом із `DepTime` і `Distance`, а потім навчіть розширену модель.
    4. Порівняйте її з моделлю без бінарного прапорця за однакового розбиття даних.
    5. Поясніть, чи справді узагальнена ознака великого аеропорту додає сигнал поверх уже наявного поля `Origin`.

- **Підказки:** якщо одночасно використовуєте `Origin` і `isMajorOrigin`, окремо прокоментуйте, чи не дублює новий прапорець уже відоме моделі знання.

---

**Варіант 15 – Балансування класів через `scale_pos_weight`:**

- **Мета:** перевірити, чи допомагає компенсація дисбалансу класів у задачі затримок рейсів.

- **Кроки:**
    1. Обчисліть співвідношення між класами `Y` і `N` у `dep_delayed_15min` та наведіть це співвідношення у звіті.
    2. Навчіть baseline-модель без класових ваг і модель з параметром `scale_pos_weight`, обчисленим із фактичного дисбалансу.
    3. Для обох моделей збережіть `precision_score`, `recall_score`, `f1_score` і confusion matrix.
    4. Порівняйте, як змінилася здатність виявляти затримані рейси та скільки при цьому зросла кількість хибних спрацьовувань.
    5. Зробіть висновок, чи виправдане використання класових ваг на ранньому етапі моделювання.

- **Підказки:** інтерпретуйте `scale_pos_weight` не лише через формулу, а й через конкретну зміну trade-off між recall та precision.

---

**Варіант 16 – `Target Encoding` для висококардинальних категорій:**

- **Мета:** порівняти цільове кодування з простішими підходами до представлення `UniqueCarrier`, `Origin` і `Dest`.

- **Кроки:**
    1. Підготуйте очищені категоріальні поля і визначте схему K-fold `target encoding`, яка не допускає використання міток поточного фолду при розрахунку середніх.
    2. Побудуйте набір ознак із `target encoding` для `UniqueCarrier`, `Origin` і `Dest`, зберігши базові числові ознаки.
    3. Як контрольну групу побудуйте альтернативний набір із `One-Hot Encoding` або `Label Encoding`.
    4. Навчіть однакові `XGBoost`-моделі та порівняйте `roc_auc_score`, `f1_score`, кількість ознак і швидкість навчання.
    5. Поясніть, чи виправдане `target encoding` саме для цього набору даних і які ризики leakage воно створює за неправильного застосування.

- **Підказки:** якщо кодуєте test set, використовуйте середні, обчислені лише на всьому очищеному train після завершення валідації.

---

**Варіант 17 – Ознака нещодавньої затримки в аеропорту відправлення:**

- **Мета:** побудувати просту проксі-ознаку ланцюгових затримок на основі попередніх рейсів.

- **Кроки:**
    1. Впорядкуйте train за `Origin`, `Month`, `DayofMonth` і `DepTime`, щоб отримати наближену часову послідовність рейсів у кожному аеропорту.
    2. Для кожного запису обчисліть `recentDelayRate` як середнє попередніх міток `dep_delayed_15min` у тому самому `Origin` у межах обраного часово-послідовного вікна.
    3. Реалізуйте розрахунок цієї ознаки без витоку інформації: для validation використовуйте лише історію з train-частини або lag-based схему всередині фолдів.
    4. Навчіть baseline-модель та модель з `recentDelayRate`, після чого порівняйте метрики й важливість нової ознаки.
    5. Поясніть, чи справді проксі ланцюгових затримок допомагає моделі, чи її сигнал поглинається базовими ознаками часу та маршруту.

- **Підказки:** не використовуйте поточну мітку рядка при побудові `recentDelayRate`; інакше отримаєте прямий leakage.

---

**Варіант 18 – Ознака аеропорту підвищеного ризику:**

- **Мета:** замінити гіпотетичні зовнішні фактори на ознаку, що повністю обчислюється з наявного train set.

- **Кроки:**
    1. На train-частині обчисліть середню частку затримок для кожного `Origin` і визначте аеропорти з верхнього квартиля за цим показником.
    2. Створіть бінарну ознаку `isHighRiskOrigin`, яка дорівнює `1` для аеропортів із підвищеним історичним ризиком затримки.
    3. Додайте цю ознаку до baseline-моделі та порівняйте її з моделлю, де використано лише `Origin` або взагалі без маршруто-залежних категорій.
    4. Наведіть порівняння якості, а також перевірте, чи нова ознака потрапляє до переліку важливих у моделі.
    5. Зробіть висновок, чи може агрегована оцінка ризику аеропорту спростити модель без помітної втрати точності.

- **Підказки:** щоб уникнути leakage, формуйте список ризикових аеропортів тільки за train-частиною кожного фолду або за окремим тренувальним піднабором.

---

**Варіант 19 – Маршрутно-спеціалізовані моделі з глобальним резервом:**

- **Мета:** оцінити, чи варто замінювати одну загальну модель набором моделей для наймасовіших маршрутів.

- **Кроки:**
    1. Створіть ознаку маршруту `Route = Origin + '_' + Dest` і визначте кілька маршрутів із найбільшою кількістю спостережень у train.
    2. Навчіть одну глобальну baseline-модель на всіх даних та окремі моделі для вибраних масових маршрутів.
    3. Для validation реалізуйте правило: якщо запис належить відомому масовому маршруту, використовуйте спеціалізовану модель, інакше – глобальну.
    4. Порівняйте гібридний підхід із суто глобальною моделлю за метрикою та окремо проаналізуйте результат на маршрутах з великою і малою кількістю спостережень.
    5. Сформулюйте висновок, чи виправдана складність route-specific підходу вже на базовому етапі лабораторної роботи.

- **Підказки:** обирайте лише ті маршрути, для яких у train є достатньо прикладів; інакше спеціалізована модель буде нестійкою.

---

**Варіант 20 – М’яке та суворе очищення даних:**

- **Мета:** порівняти наслідки мінімального і агресивного очищення для стійкості baseline-моделі.

- **Кроки:**
    1. Сформулюйте два пайплайни очищення: мінімальний, що видаляє лише явно хибні значення, і суворий, що додатково відсікає прикордонні або підозрілі спостереження.
    2. Застосуйте обидва пайплайни до train та однакові правила до test, зафіксувавши кількість збережених рядків у кожному сценарії.
    3. На кожній очищеній версії побудуйте однакову baseline-модель з незмінними ознаками та параметрами.
    4. Порівняйте метрики, частку класів і розподіл ключових ознак після різних стратегій очищення.
    5. Поясніть, чи не призводить суворе очищення до втрати корисних, хоч і рідкісних, випадків.

- **Підказки:** агресивне очищення має бути відтворюваним; не використовуйте неформальні рішення на кшталт «вилучив усе, що здалося підозрілим».

---

<a class="anchor" id="lab-10.2"></a>

## <span style="color:blue; font-size:1.5em;">10.2. Конструювання ознак та зменшення розмірності</span>

[Назад до змісту](#lab-10)


### <span style="color:red; font-size:1.5em;">Завдання 2</span>

**Для всіх варіантів:**

- Спирайтеся на очищений набір даних із Завдання 1 та використовуйте однаковий спосіб валідації, щоб зміна якості пояснювалася саме новими ознаками.
- Створюйте всі нові ознаки лише з наявних полів `Month`, `DayofMonth`, `DayOfWeek`, `DepTime`, `UniqueCarrier`, `Origin`, `Dest`, `Distance` і похідних від них.
- Якщо нова ознака базується на частоті затримок або іншій статистиці за `dep_delayed_15min`, обчислюйте її лише на тренувальній частині фолду, а для validation/test застосовуйте вже підготовлені мапінги.
- У кожному варіанті порівнюйте модель з новою ознакою проти базової конфігурації та пояснюйте, який саме механізм міг дати приріст або зниження якості.

---

**Варіант 1 – Декомпозиція `DepTime` на `DepHour` і `DepMinute`:**

- **Мета:** з’ясувати, чи дає поділ неочищеного часу вильоту на структуровані компоненти додатковий сигнал моделі.

- **Кроки:**
    1. Очистіть `DepTime`, перетворіть його на хвилини після опівночі, а потім виділіть `DepHour`, `DepMinute` та бінарний прапорець `isEarlyMorning`.
    2. Побудуйте baseline-модель із неочищеним `DepTime` і модель, у якій `DepTime` замінено набором нових часових ознак.
    3. Для обох моделей використайте той самий split і той самий набір додаткових нечасових ознак, щоб змінювався лише спосіб представлення часу.
    4. Порівняйте `roc_auc_score`, важливість ознак та за можливості побудуйте теплову карту частоти затримок або стовпчикову діаграму для `DepHour`.
    5. Поясніть, чи є структурована репрезентація часу кращою за неочищений чотирицифровий код.

- **Підказки:** `DepHour = DepTime // 100`, `DepMinute = DepTime % 100`; не забудьте відкинути або виправити неможливі значення хвилин.

---

**Варіант 2 – Зменшення розмірності після `One-Hot Encoding`:**

- **Мета:** перевірити, чи допомагає компресія високорозмірного категоріального простору після `One-Hot Encoding`.

- **Кроки:**
    1. Закодуйте `UniqueCarrier`, `Origin` і `Dest` через `One-Hot Encoding`, зберігши числові ознаки окремо.
    2. Побудуйте повний OHE-набір і альтернативний набір, у якому після кодування застосовано `PCA` або `TruncatedSVD` до фіксованої кількості компонент.
    3. Навчіть `XGBoost` на обох представленнях, контролюючи однаковий split, метрику і random seed.
    4. Порівняйте якість, кількість ознак після стиснення та зміну часу навчання моделі.
    5. Зробіть висновок, чи виправдане зменшення розмірності для цього набору даних і чи не втрачається при цьому критичний сигнал.

- **Підказки:** якщо матриця після OHE розріджена, віддайте перевагу `TruncatedSVD`, а не класичному `PCA`.

---

**Варіант 3 – `weekend_indicator` і `peak_travel_indicator`:**

- **Мета:** оцінити внесок простих календарних прапорців, пов’язаних із поведінкою пасажиропотоку.

- **Кроки:**
    1. Очистіть `DayOfWeek` і `DayofMonth`, після чого створіть `weekend_indicator` та інший індикатор пікових дат місяця, наприклад `peak_travel_indicator`.
    2. Проаналізуйте частку затримок для кожного з нових прапорців у train і перевірте, чи відрізняється вона від середнього рівня.
    3. Побудуйте baseline-модель без нових ознак і розширену модель з двома бінарними прапорцями.
    4. Порівняйте метрики і перевірте, чи потрапили нові ознаки до числа важливих у моделі.
    5. Поясніть, чи можуть такі прості календарні прапорці замінити складніші часові ознаки.

- **Підказки:** перед створенням `weekend_indicator` перевірте, як саме кодуються дні тижня у вашому очищеному датасеті.

---

**Варіант 4 – Бінування `Distance` та крос-ознаки з часом доби:**

- **Мета:** перевірити, чи додає інтерпретована категоризація відстані корисний сигнал моделі.

- **Кроки:**
    1. Розбийте `Distance` на кілька осмислених інтервалів, наприклад `short-haul`, `medium-haul`, `long-haul`, і зафіксуйте пороги.
    2. На основі `DepHour` сформуйте широкі часові блоки доби та створіть крос-ознаку на кшталт `distance_bin + daypart`.
    3. Підготуйте три набори ознак: із неочищеною `Distance`, лише з бінованою `Distance` та з бінованою `Distance` плюс крос-ознака.
    4. Навчіть моделі для кожного набору і порівняйте `roc_auc_score`, інтерпретованість та важливість нових категорій.
    5. Зробіть висновок, чи спрощене представлення відстані працює краще за безперервну величину.

- **Підказки:** якщо створюєте крос-ознаку, закодуйте її як категоріальну, а не як штучний числовий індекс.

---

**Варіант 5 – Циклічне кодування `Month` і `DayOfWeek`:**

- **Мета:** перевірити, чи допомагає циклічне представлення періодичних календарних ознак.

- **Кроки:**
    1. Перетворіть `Month` і `DayOfWeek` у числовий формат та підготуйте для них як звичайне, так і циклічне представлення через `sin` та `cos`.
    2. Зберіть принаймні дві моделі: одну зі звичайними календарними числами або OHE, а іншу – з циклічними перетвореннями.
    3. Навчіть `XGBoost` за однакових умов і порівняйте результат за обраною метрикою.
    4. Перевірте, чи входять циклічні ознаки до числа важливих і чи є вони стабільними в різних фолдах.
    5. Поясніть, чи відображає циклічне кодування повторювану природу календаря краще за прямі числові значення.

- **Підказки:** для дня тижня використайте формули виду `sin(2π * DayOfWeek / 7)` і `cos(2π * DayOfWeek / 7)`.

---

**Варіант 6 – Маршрутна ознака `Route = Origin + Dest`:**

- **Мета:** перевірити, чи є поєднання `Origin` та `Dest` інформативнішим за їх окреме використання.

- **Кроки:**
    1. Створіть ознаку `Route`, об’єднавши `Origin` і `Dest`, та проаналізуйте кількість унікальних маршрутів і найчастіші значення.
    2. Побудуйте один набір ознак із роздільними `Origin` і `Dest`, а другий – з однією маршрутною ознакою `Route` та числовими полями.
    3. Закодуйте категорії однаковим способом у межах кожного сценарію і навчіть дві порівнювані `XGBoost`-моделі.
    4. Порівняйте якість, кількість ознак і важливість маршруту відносно окремих аеропортів.
    5. Зробіть висновок, чи варто використовувати маршрут як окрему сутність у задачі прогнозування затримок.

- **Підказки:** якщо кількість унікальних маршрутів дуже велика, додатково проаналізуйте варіант із об’єднанням рідкісних маршрутів у `OtherRoute`.

---

**Варіант 7 – Рангова трансформація числових ознак:**

- **Мета:** перевірити, чи поліпшує перехід до процентильних рангів роботу моделі за наявності перекошених розподілів.

- **Кроки:**
    1. Для `Distance` і, за потреби, `DepTime` побудуйте розподіли та перевірте, чи містять вони важкі хвости або значну асиметрію.
    2. Створіть rank-based представлення числових ознак через `rank(pct=True)` на train-частині, а для validation/test застосуйте відповідні мапінги або узгоджений спосіб перетворення.
    3. Підготуйте три моделі: з неочищеними числовими ознаками, зі стандартним масштабуванням і з рангами.
    4. Порівняйте `f1_score`, `precision`, `recall` та поясніть, чи зменшився вплив екстремальних значень.
    5. Зробіть висновок, чи є рангова трансформація виправданою саме для задачі затримок рейсів.

- **Підказки:** не змішуйте розрахунок рангу на train і validation; інакше перетворення втратить коректність.

---

**Варіант 8 – Ознака співвідношення часу вильоту та відстані:**

- **Мета:** перевірити, чи містить комбінована числова характеристика `DepTime` і `Distance` додатковий сигнал.

- **Кроки:**
    1. Очистіть `DepTime`, переведіть його у хвилини після опівночі й підготуйте коректну числову версію `Distance`.
    2. Створіть нову ознаку, наприклад `DepTimeDistanceRatio = DepMinutes / (Distance + 1)`, і коротко обґрунтуйте її інтерпретацію.
    3. Навчіть baseline-модель без цієї ознаки та модель з доданим співвідношенням.
    4. Порівняйте метрики і перевірте, чи входить нова ознака до числа важливих у моделі.
    5. Поясніть, чи є така комбінована характеристика справді змістовною, чи лише дублює вихідні числові ознаки.

- **Підказки:** якщо співвідношення має дуже перекошений розподіл, додатково перевірте логарифмування або clipping.

---

**Варіант 9 – Приблизний `day_of_year` як інтегральна сезонна ознака:**

- **Мета:** дослідити, чи краще працює інтегрована сезонна шкала, ніж окремі `Month` і `DayofMonth`.

- **Кроки:**
    1. Після очищення календарних полів побудуйте наближену ознаку `day_of_year`, що відображає порядковий день у році.
    2. Залиште контрольну модель з окремими `Month` і `DayofMonth`, а для другої моделі замініть їх на `day_of_year`.
    3. Навчіть обидві моделі за однакових умов та оцініть їх за спільною метрикою.
    4. Порівняйте, як змінилася важливість сезонних ознак і чи не втратилися локальні місячні ефекти.
    5. Зробіть висновок, чи придатна спрощена сезонна координата для цієї задачі.

- **Підказки:** якщо використовуєте наближений календар, прямо зазначте, що це модельна апроксимація без реального календарного перерахунку дат.

---

**Варіант 10 – Поліноміальні часові ознаки і перевірка надлишковості:**

- **Мета:** перевірити, чи дають квадратичні та перехресні часові терміни реальний приріст моделі.

- **Кроки:**
    1. Сформуйте базові часові ознаки `DepHour`, `DepMinute` та, за потреби, `Distance`, а також створіть для них поліноміальні перетворення другого ступеня.
    2. До навчання перевірте кореляції між новими ознаками, щоб не включати очевидно дубльовані або майже тотожні величини.
    3. Побудуйте модель на базовому часовому наборі й окремо модель на розширеному поліноміальному просторі.
    4. Порівняйте якість, час навчання та зміну важливості ознак після розширення простору.
    5. Поясніть, чи справді поліноміальні часові терміни додають інформацію поверх уже нелінійної моделі `XGBoost`.

- **Підказки:** для перевірки надлишковості достатньо матриці кореляцій і короткого коментаря про найсильніше пов’язані пари ознак.

---

**Варіант 11 – Кластерні типи рейсів як нова категоріальна ознака:**

- **Мета:** перетворити схожі рейси на компактні «типи», придатні для подальшого моделювання.

- **Кроки:**
    1. Виберіть кілька числових ознак, наприклад `DepHour`, `DepMinute` і `Distance`, та стандартизуйте їх для кластеризації.
    2. Виконайте `KMeans` або інший простий метод кластеризації, підібравши розумну кількість кластерів і коротко обґрунтувавши її.
    3. Додайте номер кластера як нову категоріальну ознаку до baseline-набору ознак.
    4. Навчіть моделі без кластерної ознаки та з нею, після чого порівняйте якість і затримкові профілі отриманих кластерів.
    5. Поясніть, чи можна інтерпретувати кластери як окремі типи рейсів із різним ризиком затримки.

- **Підказки:** не змішуйте етап кластеризації з міткою `dep_delayed_15min`; це має бути суто unsupervised-ознака.

---

**Варіант 12 – Відбір ознак через `mutual information` або `chi-square`:**

- **Мета:** перевірити, чи можна скоротити простір ознак без суттєвої втрати якості.

- **Кроки:**
    1. Сформуйте розширений набір ознак із календарними, маршрутними, числовими та похідними полями, який стане кандидатом для відбору.
    2. Застосуйте `mutual_info_classif`, `chi2` або інший придатний критерій для ранжування ознак відносно `dep_delayed_15min`.
    3. Виберіть кілька сценаріїв – наприклад top-10, top-20 та повний набір ознак – і підготуйте відповідні матриці даних.
    4. Навчіть `XGBoost` на кожному сценарії та порівняйте якість, швидкість навчання і компактність моделі.
    5. Зробіть висновок, чи здатен формальний відбір ознак зберегти основний сигнал у цій задачі.

- **Підказки:** не порівнюйте критерії відбору на різних сплітах; усі сценарії мають будуватися на одному train/validation-поділі.

---

**Варіант 13 – Оцінка завантаженості аеропорту в часовому вікні:**

- **Мета:** перетворити локальну щільність рейсів на ознаку операційного навантаження.

- **Кроки:**
    1. Переведіть `DepTime` у хвилини після опівночі та для кожного `Origin` відсортуйте рейси всередині дня.
    2. Для кожного запису порахуйте кількість рейсів із того самого `Origin`, що відправляються у вікні, наприклад, `±120` хвилин.
    3. Додайте нову ознаку `originCongestionScore` до базового набору та навчіть модель з нею.
    4. Порівняйте якість із baseline і перевірте, чи зростає частка затримок у групах з високою локальною завантаженістю.
    5. Поясніть, чи є локальна щільність вильотів корисною проксі-ознакою для затримок.

- **Підказки:** якщо повний підрахунок у вікні надто дорогий обчислювально, використайте бінування часу за годинами та оцініть наближену щільність.

---

**Варіант 14 – Сегменти тижня як агрегована календарна ознака:**

- **Мета:** перевірити, чи краще працює укрупнене групування днів тижня, ніж неочищені значення `DayOfWeek`.

- **Кроки:**
    1. Побудуйте мапінг, який перетворює `DayOfWeek` на кілька сегментів, наприклад `Mon–Wed`, `Thu–Fri`, `Weekend`.
    2. Порівняйте частку затримок у кожному сегменті та дослідіть, як вони взаємодіють із `DepHour` або `Distance`.
    3. Підготуйте модель із неочищеним `DayOfWeek` та окремо модель із новою укрупненою ознакою `weekSegment`.
    4. Порівняйте якість, важливість ознак і простоту інтерпретації між двома сценаріями.
    5. Зробіть висновок, чи є сегментація тижня корисним компромісом між простотою і якістю.

- **Підказки:** мапінг має бути осмисленим; не створюйте довільні групи днів без пояснення логіки поділу.

---

**Варіант 15 – Проксі-ознака ймовірності пересадкового навантаження:**

- **Мета:** створити виконуваний аналог ознаки «ймовірна пересадка» лише з наявних полів.

- **Кроки:**
    1. Для кожного `Origin` і часової точки оцініть інтенсивність відправлень до різних `Dest` у вікні 90–120 хвилин навколо `DepTime`.
    2. На основі цієї щільності сформуйте бінарну або рангову ознаку `isLikelyConnection`, яка позначає рейси з підвищеним пересадковим навантаженням.
    3. Додайте нову ознаку до базового набору й навчіть модель, не змінюючи інших перетворень.
    4. Порівняйте якість із baseline, а також окремо перевірте розподіл затримок для груп з низькою та високою ймовірністю пересадкової активності.
    5. Поясніть, чи є така проксі-ознака корисною заміною недоступних реальних даних про стикування пасажирів.

- **Підказки:** у звіті прямо підкресліть, що це не реальна ознака пасажирських пересадок, а лише операційне наближення на основі щільності розкладу.

---

**Варіант 16 – Мета-ознака історичної частоти затримок за годиною:**

- **Мета:** перевірити, чи корисніше подати модельований ризик години вильоту, ніж неочищений часовий код.

- **Кроки:**
    1. Перетворіть `DepTime` на `DepHour` і для train-частини кожного фолду обчисліть середню частку затримок окремо для кожної години.
    2. Зіставте кожному запису ознаку `hourDelayRate`, отриману лише з тренувальної частини відповідного фолду.
    3. Побудуйте baseline-модель із неочищеним `DepTime` або `DepHour` та модель з додатковою мапованою ознакою `hourDelayRate`.
    4. Порівняйте метрики і перевірте, чи не виявився статистичний мета-сигнал сильнішим за початкове числове представлення часу.
    5. Зробіть висновок, чи виправдане таке статистичне узагальнення у задачі затримок рейсів.

- **Підказки:** застосовуйте fold-based обчислення, інакше `hourDelayRate` міститиме leakage з validation.

---

**Варіант 17 – Відхилення дальності маршруту від середньої для `Origin`:**

- **Мета:** використати відносне, а не абсолютне значення `Distance` для характеристики рейсу.

- **Кроки:**
    1. Для кожного `Origin` обчисліть середнє значення `Distance` на train-частині та побудуйте довідник середніх відстаней.
    2. Створіть нову ознаку `distanceDeviation = Distance - meanDistanceByOrigin`.
    3. Додайте цю ознаку до baseline-набору й побудуйте модель з та без `distanceDeviation`.
    4. Порівняйте якість і перевірте, чи відрізняється частка затримок для рейсів з великим додатним або від’ємним відхиленням.
    5. Поясніть, чи є відносна дальність маршруту інформативнішою за абсолютну.

- **Підказки:** якщо певний `Origin` трапляється лише в validation/test, використайте глобальне середнє або окремо зафіксуйте резервне правило.

---

**Варіант 18 – `weekendRatio` для аеропорту відправлення:**

- **Мета:** створити агреговану ознаку структури тижневого потоку рейсів для кожного `Origin`.

- **Кроки:**
    1. Для кожного аеропорту в train порахуйте кількість рейсів у вихідні та будні, а потім сформуйте `weekendRatio`.
    2. Приєднайте цей показник до кожного рядка за ключем `Origin`, отримавши нову числову ознаку.
    3. Побудуйте baseline-модель і модель з `weekendRatio`, не змінюючи інших ознак.
    4. Порівняйте метрики та перевірте, чи корелює високий `weekendRatio` із підвищеною або зниженою часткою затримок.
    5. Зробіть висновок, чи містить тижнева структура потоку аеропорту самостійну прогностичну цінність.

- **Підказки:** у формулі використайте невелике згладжування в знаменнику, щоб уникнути ділення на нуль.

---

**Варіант 19 – Порядковий номер рейсу в добовому розкладі аеропорту:**

- **Мета:** перевірити, чи накопичується ризик затримки для пізніших рейсів у межах дня.

- **Кроки:**
    1. Відсортуйте рейси за `Origin`, `Month`, `DayofMonth` і `DepTime`, а потім присвойте кожному рядку порядковий номер `flightIndex` у межах дня для конкретного аеропорту.
    2. Побудуйте описову статистику `flightIndex` і перевірте, як змінюється частка затримок для ранніх і пізніх позицій у розкладі.
    3. Додайте `flightIndex` до baseline-набору ознак і навчіть розширену модель.
    4. Порівняйте її з моделлю, що використовує лише `DepTime`, та оцініть, яка з двох часових ознак виявляється сильнішою.
    5. Зробіть висновок, чи несе порядкова позиція рейсу в добовому потоці інформацію, відсутню у неочищеному часі вильоту.

- **Підказки:** у разі однакового `DepTime` порядок можна розв’язувати стабільним сортуванням за індексом рядка.

---

**Варіант 20 – Композитні блоки `DayOfWeek × DepHour`:**

- **Мета:** перевірити, чи допомагає злиття дня тижня і широкого часу доби в одну категорію.

- **Кроки:**
    1. Побудуйте спрощені категорії для днів тижня, наприклад `Weekday` і `Weekend`, а також широкі часові блоки за `DepHour`.
    2. Об’єднайте їх у нову категорію на зразок `weekday_morning`, `weekday_evening`, `weekend_night`.
    3. Закодуйте композитну ознаку й навчіть модель з нею, а як контроль використайте модель з окремими `DayOfWeek` і `DepHour`.
    4. Порівняйте якість, важливість нової композитної ознаки та зміну кількості параметрів після кодування.
    5. Поясніть, чи справді об’єднаний часово-календарний блок краще відображає синергію між днем і часом вильоту.

- **Підказки:** не робіть надто дрібні часові сегменти; для початку достатньо 3–4 інтервалів доби.

---

<a class="anchor" id="lab-10.3"></a>

## <span style="color:blue; font-size:1.5em;">10.3. Налаштування гіперпараметрів та оцінювання моделі</span>

[Назад до змісту](#lab-10)


### <span style="color:red; font-size:1.5em;">Завдання 3</span>

**Для всіх варіантів:**

- Використовуйте один і той самий очищений набір ознак та незмінний спосіб валідації, щоб усі порівняння гіперпараметрів були коректними.
- Фіксуйте `random_state`, перелік параметрів, діапазон пошуку, обрану метрику (`roc_auc_score`, `f1_score`, `accuracy_score`) і час навчання.
- Під час tuning не використовуйте `flight_delays_test.csv` для оцінювання якості; test set дозволено застосовувати лише для узгоджених перетворень і фінального прогнозу після завершення валідації.
- У висновку кожного варіанта вкажіть найкращу конфігурацію, поясніть компроміс між якістю і складністю та коротко охарактеризуйте стабільність результату.

---

**Варіант 1 – `GridSearchCV` для `max_depth` і `n_estimators`:**

- **Мета:** отримати базову карту впливу двох найважливіших параметрів складності `XGBoost`.

- **Кроки:**
    1. Зафіксуйте очищений набір ознак і підготуйте сітку значень для `max_depth` та `n_estimators`, що охоплює малі, середні й великі моделі.
    2. Запустіть `GridSearchCV` або еквівалентний ручний перебір з однаковою схемою крос-валідації для всіх комбінацій.
    3. Для кожної комбінації збережіть середнє значення метрики й час навчання, а потім визначте найкращу конфігурацію.
    4. Перенавчіть модель з найкращими параметрами на train-частині та перевірте її на validation для додаткового контролю.
    5. Поясніть, як зміна глибини і кількості дерев вплинула на точність, переобладнання та обчислювальну вартість.

- **Підказки:** не змінюйте водночас `learning_rate`, щоб оцінка впливу `max_depth` і `n_estimators` залишалася чистою.

---

**Варіант 2 – Аналіз впливу `learning_rate`:**

- **Мета:** дослідити компроміс між швидкістю збіжності та ризиком переобладнання.

- **Кроки:**
    1. Зафіксуйте всі параметри моделі, окрім `learning_rate`, і сформуйте послідовність значень від низьких до високих.
    2. Навчіть `XGBoost` для кожного значення `learning_rate`, використавши однакову кількість дерев та однаковий split.
    3. Збережіть train- і validation-метрики, а також час навчання для кожного запуску.
    4. Побудуйте таблицю або графік залежності якості від `learning_rate` і визначте діапазон, у якому модель працює найкраще.
    5. Зробіть висновок, коли висока швидкість навчання починає шкодити узагальненню, а надто мала – лише збільшує обчислювальні витрати.

- **Підказки:** якщо якість зростає при малому `learning_rate`, перевірте, чи достатньо велике значення `n_estimators` для чесного порівняння.

---

**Варіант 3 – Рання зупинка `early_stopping_rounds`:**

- **Мета:** автоматично підібрати ефективну кількість бустингових ітерацій без повного перебору `n_estimators`.

- **Кроки:**
    1. Підготуйте окрему validation-частину та задайте для `XGBoost` досить велике значення `n_estimators`, наприклад 800–1000.
    2. Увімкніть `early_stopping_rounds` і навчіть модель з відстеженням валідаційної метрики на кожній ітерації.
    3. Зафіксуйте найкращу ітерацію, значення метрики на ній і фактичну кількість використаних дерев.
    4. Порівняйте підсумкову якість і час навчання з моделлю, яка відпрацьовує повний фіксований бюджет ітерацій.
    5. Поясніть, чи дала рання зупинка економію часу без втрати якості та наскільки стабільним є знайдений оптимум.

- **Підказки:** слідкуйте, щоб рання зупинка відбувалася за тією ж метрикою, яку ви потім використовуєте для основного порівняння.

---

**Варіант 4 – Перебір `subsample` і `colsample_bytree`:**

- **Мета:** перевірити, чи зменшують вибіркові підвибірки рядків і ознак переобладнання моделі.

- **Кроки:**
    1. Зафіксуйте базову конфігурацію `XGBoost` і визначте сітку значень для `subsample` та `colsample_bytree`.
    2. Запустіть повний або частковий перебір комбінацій, використавши однакову крос-валідацію і метрику.
    3. Для кожної конфігурації збережіть середній результат, стандартне відхилення та, за можливості, час навчання.
    4. Побудуйте зведену таблицю або теплову карту, що покаже, які поєднання параметрів працюють найкраще.
    5. Поясніть, чи справді випадковий відбір рядків і ознак підвищує узагальнювальну здатність моделі на цьому наборі даних.

- **Підказки:** не змінюйте одночасно інші параметри складності, інакше інтерпретація впливу sampling-параметрів втратить чіткість.

---

**Варіант 5 – Підбір `scale_pos_weight` для дисбалансованої цілі:**

- **Мета:** оцінити, як зміна ваги позитивного класу впливає на виявлення затриманих рейсів.

- **Кроки:**
    1. Обчисліть фактичне співвідношення класів у `dep_delayed_15min` і підготуйте кілька кандидатів для `scale_pos_weight`, включно з емпіричним значенням.
    2. Навчіть низку моделей із різними вагами позитивного класу, не змінюючи інших параметрів.
    3. Для кожної моделі порівняйте `precision`, `recall`, `f1_score` і confusion matrix на validation.
    4. Визначте конфігурацію, яка дає найкращий компроміс між знаходженням затриманих рейсів і кількістю хибних тривог.
    5. Поясніть, чи виправдане посилення ваги класу `Y` для вашого способу оцінювання.

- **Підказки:** цей варіант слід оцінювати не лише за `accuracy_score`, бо точність може маскувати погану якість на міноритарному класі.

---

**Варіант 6 – `RandomizedSearchCV` у широкому просторі параметрів:**

- **Мета:** порівняти випадковий пошук із повним перебором у задачі tuning `XGBoost`.

- **Кроки:**
    1. Визначте широкий, але реалістичний простір для `max_depth`, `learning_rate`, `subsample`, `colsample_bytree`, `min_child_weight` та `gamma`.
    2. Запустіть `RandomizedSearchCV` з обмеженим бюджетом ітерацій, зафіксувавши `random_state` та метрику.
    3. Виберіть найкращу випадково знайдену конфігурацію і перенавчіть модель на train-частині для перевірки на validation.
    4. Порівняйте результат із коротким `GridSearchCV` або з ручно підібраною baseline-конфігурацією.
    5. Зробіть висновок, чи є випадковий пошук ефективним компромісом між якістю та обчислювальними витратами.

- **Підказки:** не робіть простір пошуку надто вузьким; інакше перевага випадкового пошуку над сіткою буде штучно заниженою.

---

**Варіант 7 – Послідовна «байєсівська» оптимізація параметрів:**

- **Мета:** перевірити, чи дає послідовний інтелектуальний пошук параметрів виграш за однакового бюджету випробувань.

- **Кроки:**
    1. Підготуйте той самий очищений набір ознак і зафіксуйте бюджет оцінювань, який буде однаковим для послідовного пошуку та `RandomizedSearchCV`.
    2. Використайте `BayesSearchCV`, `Optuna` або інший послідовний оптимізатор; якщо пакет відсутній, відтворіть ідею через покрокове ручне звуження діапазонів.
    3. Логіруйте запропоновані конфігурації, значення метрики та номер ітерації, щоб було видно динаміку пошуку.
    4. Порівняйте найкращу знайдену конфігурацію та час пошуку з результатом випадкового або сіткового підходу.
    5. Зробіть висновок, чи виправдана складність послідовної оптимізації для цієї лабораторної роботи.

- **Підказки:** у звіті чітко відокремте результат оптимізатора від вашої фінальної перевірки моделі на validation.

---

**Варіант 8 – Порівняння кількох метрик якості:**

- **Мета:** показати, що одна й та сама модель може виглядати по-різному залежно від обраної метрики.

- **Кроки:**
    1. Зафіксуйте одну помірно налаштовану `XGBoost`-модель і навчіть її на очищеному наборі ознак.
    2. На validation обчисліть `accuracy_score`, `precision_score`, `recall_score`, `f1_score` і `roc_auc_score`.
    3. Подайте результати у зведеній таблиці та окремо побудуйте confusion matrix для контексту.
    4. Порівняйте, які метрики дають найбільш оптимістичну й найбільш вимогливу оцінку моделі.
    5. Обґрунтуйте, яку метрику варто вважати головною саме для задачі прогнозування затримок вильоту.

- **Підказки:** у висновку не обмежуйтеся твердженням «метрики різні»; поясніть, яке рішення може бути хибним, якщо орієнтуватися лише на `accuracy`.

---

**Варіант 9 – Time-based cross-validation за місяцями:**

- **Мета:** перевірити tuning моделі в умовах, наближених до прогнозування майбутніх періодів.

- **Кроки:**
    1. Після очищення календарних полів сформуйте часові фолди так, щоб train містив ранні місяці, а validation – пізніші.
    2. На цих time-based split-ах виконайте добір принаймні двох параметрів, наприклад `max_depth` і `learning_rate`.
    3. Окремо проведіть аналогічний добір на звичайній випадковій крос-валідації з тим самим набором кандидатів.
    4. Порівняйте найкращі конфігурації та отримані оцінки якості в обох сценаріях.
    5. Поясніть, чому параметри, оптимальні для часової валідації, можуть відрізнятися від параметрів, знайдених на випадкових фолдах.

- **Підказки:** у зведеній таблиці бажано показати не лише найкращий результат, а й розкид якості між часовими фолдами.

---

**Варіант 10 – `validation_curve` для `max_depth`:**

- **Мета:** простежити, як глибина дерев впливає на недонавчання і переобладнання.

- **Кроки:**
    1. Зафіксуйте всі параметри моделі, крім `max_depth`, та сформуйте послідовність значень глибини від дуже малої до помірно великої.
    2. Для кожного значення глибини обчисліть train- і validation-оцінки за допомогою `validation_curve` або еквівалентного ручного циклу.
    3. Побудуйте графік двох кривих, що відображає зміну якості на train і validation зі зростанням `max_depth`.
    4. Визначте область, де модель переходить від недонавчання до переобладнання, і зіставте це з числовими результатами.
    5. Зробіть висновок, яке значення `max_depth` є найбільш виправданим для цього набору даних.

- **Підказки:** інтерпретуючи криві, звертайте увагу не лише на максимум validation-метрики, а й на розрив між train і validation.

---

**Варіант 11 – Компроміс між часом навчання і якістю:**

- **Мета:** підібрати конфігурацію, яка є не лише точною, а й помірною за обчислювальними витратами.

- **Кроки:**
    1. Складіть набір конфігурацій з різними поєднаннями `max_depth` і `n_estimators`, що охоплюють легкі та важкі моделі.
    2. Для кожної конфігурації виміряйте час навчання та validation-метрику в однакових умовах.
    3. Подайте результат у вигляді таблиці або точкової діаграми `час навчання – якість`.
    4. Виділіть конфігурації, що дають найкращий компроміс, і окремо покажіть точки з явним ефектом diminishing returns.
    5. Поясніть, яку конфігурацію ви б обрали для практичної роботи студента і чому.

- **Підказки:** вимірюйте лише час безпосереднього навчання; інакше сторонні витрати на підготовку даних спотворять висновок про складність моделі.

---

**Варіант 12 – Ансамбль кількох добре налаштованих `XGBoost`-моделей:**

- **Мета:** перевірити, чи дає усереднення сильних, але різних конфігурацій приріст якості.

- **Кроки:**
    1. Підберіть кілька конкурентних конфігурацій `XGBoost`, які показують близьку якість, але різняться структурою параметрів.
    2. Навчіть кожну модель окремо та отримайте прогноз імовірностей на спільній validation-вибірці.
    3. Побудуйте ансамбль простим середнім або зваженим середнім прогнозів і зафіксуйте правило обчислення ваг.
    4. Порівняйте ансамбль з найкращою окремою моделлю за `roc_auc_score`, `f1_score` і стабільністю результату.
    5. Зробіть висновок, чи виправданий ансамбль із кількох версій одного й того самого алгоритму.

- **Підказки:** для чесного порівняння беріть моделі, що дійсно відрізняються, а не три майже ідентичні конфігурації.

---

**Варіант 13 – Nested cross-validation для tuning і оцінювання:**

- **Мета:** отримати менш упереджену оцінку якості після добору гіперпараметрів.

- **Кроки:**
    1. Побудуйте зовнішній цикл крос-валідації, який буде відповідати за підсумкову оцінку моделі.
    2. Усередині кожного зовнішнього фолду запускайте внутрішній пошук гіперпараметрів через `GridSearchCV` або `RandomizedSearchCV`.
    3. Для кожного зовнішнього фолду збережіть найкращі параметри, validation-метрику та різницю між внутрішньою і зовнішньою оцінками.
    4. Узагальніть результати по зовнішніх фолдах, обчисливши середнє та стандартне відхилення якості.
    5. Поясніть, чому nested CV дає надійнішу оцінку, але водночас суттєво збільшує обчислювальні витрати.

- **Підказки:** у висновку корисно порівняти nested CV з простим одноетапним tuning, щоб показати реальну різницю в оцінках.

---

**Варіант 14 – Аналіз параметра `gamma`:**

- **Мета:** дослідити, як мінімальне допустиме зменшення втрат на розщепленні впливає на складність дерева.

- **Кроки:**
    1. Зафіксуйте базову конфігурацію моделі та сформуйте послідовність значень `gamma` від `0` до помітно більшого обмеження.
    2. Навчіть модель для кожного значення параметра, зберігаючи validation-метрику, час навчання та інші релевантні характеристики.
    3. Порівняйте, як змінюються якість і стабільність моделі зі зростанням `gamma`.
    4. За можливості доповніть аналіз описом зміни складності моделі або непрямими ознаками, наприклад зменшенням різниці між train і validation.
    5. Зробіть висновок, чи допомагає `gamma` стримувати переобладнання саме у вашому випадку.

- **Підказки:** варто інтерпретувати `gamma` не ізольовано, а разом із глибиною дерева та кількістю дерев у моделі.

---

**Варіант 15 – Регуляризація `reg_alpha` і `reg_lambda`:**

- **Мета:** перевірити, чи покращує регуляризація узагальнення бустингової моделі.

- **Кроки:**
    1. Зафіксуйте baseline-конфігурацію та підготуйте окремі сценарії: без регуляризації, з `reg_alpha` і з `reg_lambda`.
    2. Навчіть моделі на однаковому split і зафіксуйте validation-метрики, а також поведінку на train.
    3. Побудуйте зведену таблицю, де буде видно зміну якості під впливом кожного типу регуляризації.
    4. Оцініть, чи зменшився розрив між train і validation після введення штрафів.
    5. Поясніть, який тип регуляризації є доцільнішим для цієї задачі і чому.

- **Підказки:** не вводьте великі штрафи одразу; почніть із помірних значень, щоб не придушити модель надмірно.

---

**Варіант 16 – `RepeatedKFold` для стабільнішої оцінки якості:**

- **Мета:** з’ясувати, наскільки результат моделі залежить від випадкового вибору фолдів.

- **Кроки:**
    1. Зафіксуйте одну помірно налаштовану `XGBoost`-модель і застосуйте до неї `RepeatedKFold` з кількома повторами.
    2. Для кожного фолду і повтору збережіть значення основної метрики, а потім обчисліть середнє та стандартне відхилення.
    3. Окремо побудуйте звичайну 5-fold-оцінку без повторів для порівняння.
    4. Порівняйте стабільність оцінок у двох сценаріях і визначте, чи змінюється висновок про якість моделі.
    5. Поясніть, коли повторна крос-валідація виправдана в навчальному експерименті.

- **Підказки:** не змінюйте набір ознак під час цього порівняння; інакше розкид буде пояснюватися не лише схемою CV.

---

**Варіант 17 – Інтерпретація tuned-моделі через `PDP` або `SHAP`:**

- **Мета:** поєднати підбір параметрів з аналізом впливу ключових ознак на прогноз.

- **Кроки:**
    1. Підберіть або візьміть уже знайдену сильну конфігурацію `XGBoost` і навчіть її на очищеному наборі ознак.
    2. Виберіть кілька важливих ознак, наприклад `Distance`, `DepTime` або `UniqueCarrier`, та побудуйте `PDP` або `SHAP`-графіки.
    3. Порівняйте, чи узгоджується інтерпретація моделі з вашими очікуваннями щодо ризику затримок.
    4. У звіті зіставте числову якість tuned-моделі з якісними висновками з графіків інтерпретації.
    5. Поясніть, чому інтерпретація важлива навіть тоді, коли tuning уже дав хороший результат за метрикою.

- **Підказки:** якщо використовуєте `SHAP`, покажіть принаймні один глобальний і один локальний тип візуалізації.

---

**Варіант 18 – Логування метрики на кожній ітерації boosting:**

- **Мета:** простежити, як змінюється якість моделі в процесі нарощування дерев.

- **Кроки:**
    1. Підготуйте train/validation split і навчіть `XGBoost` з `eval_set`, увімкнувши логування обраної метрики на ітераціях.
    2. Збережіть історію метрики для train і validation та перетворіть її на криву навчання.
    3. Визначте момент, коли validation-метрика перестає зростати або починає коливатися без покращення.
    4. Порівняйте побачену динаміку з параметрами `n_estimators` і `early_stopping_rounds`, якщо ви їх використовуєте.
    5. Зробіть висновок, чи досягає модель насичення рано, чи їй потрібен довший бустинговий процес.

- **Підказки:** для наочності варто підписати на графіку найкращу ітерацію та відповідне значення метрики.

---

**Варіант 19 – Чутливість моделі до `random_state`:**

- **Мета:** оцінити, наскільки результат tuning залежить від випадкових компонентів навчання.

- **Кроки:**
    1. Візьміть одну помірно налаштовану конфігурацію `XGBoost` і сформуйте перелік кількох значень `random_state`.
    2. Навчіть модель окремо для кожного значення seed за однакового split і незмінних ознак.
    3. Збережіть метрики для кожного запуску та обчисліть середнє і стандартне відхилення.
    4. Порівняйте, чи змінюється ранжування ознак або лише кінцева метрика коливається в допустимих межах.
    5. Поясніть, чи суттєво впливає випадковість на висновки щодо якості моделі в цій задачі.

- **Підказки:** якщо використовуєте sampling-параметри типу `subsample`, вплив `random_state` зазвичай стає помітнішим.

---

**Варіант 20 – Локальний аналіз чутливості гіперпараметрів:**

- **Мета:** зрозуміти, які параметри tuned-моделі є найбільш критичними до невеликих змін.

- **Кроки:**
    1. Виберіть одну сильну базову конфігурацію та перелік параметрів для локального аналізу, наприклад `max_depth`, `learning_rate`, `subsample`, `colsample_bytree`.
    2. Змінюйте кожен параметр окремо на невеликий крок навколо базового значення, не торкаючись решти конфігурації.
    3. Для кожної модифікації обчисліть validation-метрику та зафіксуйте відхилення від базового результату.
    4. Побудуйте короткий рейтинг параметрів за силою впливу на якість моделі.
    5. Зробіть висновок, які параметри варто добирати найретельніше, а які є відносно нечутливими у цій задачі.

- **Підказки:** локальна чутливість не замінює повний tuning, але добре показує, де модель найбільш вразлива до неточного налаштування.

---

<a class="anchor" id="lab-10.4"></a>

## <span style="color:blue; font-size:1.5em;">10.4. Обробка пропущених даних, викидів та особливих випадків</span>

[Назад до змісту](#lab-10)


### <span style="color:red; font-size:1.5em;">Завдання 4</span>

**Для всіх варіантів:**

- Усі експерименти з пропусками, шумом, викидами або синтетичним пошкодженням виконуйте на копії навчальних даних, щоб мати чисту контрольну версію для порівняння.
- Якщо ви навмисно вводите пропуски чи корупцію, фіксуйте частку змінених рядків, random seed і правило модифікації, щоб експеримент можна було відтворити.
- Будь-яку узгоджену функцію очищення застосовуйте однаково до `train` і `test`, але оцінюйте якість лише там, де доступна `dep_delayed_15min`.
- У звіті окремо подавайте зміну метрики, кількість змінених рядків, новий розподіл ознак і короткий коментар про те, чи придатний метод для реального пайплайна підготовки даних.

---

**Варіант 1 – Штучні пропуски в `DepTime`:**

- **Мета:** порівняти видалення та імпутацію в умовах контрольовано неповного часу вильоту.

- **Кроки:**
    1. Візьміть очищений train і випадково зробіть відсутніми значення `DepTime` для фіксованої частки рядків, зберігши список змінених індексів.
    2. Реалізуйте дві стратегії: вилучення рядків з пропусками та заповнення `DepTime` глобальною медіаною або середнім значенням.
    3. Побудуйте окремі `XGBoost`-моделі для кожної стратегії на однаковому split.
    4. Порівняйте `f1_score`, `roc_auc_score`, кількість збережених спостережень і зміну розподілу часу вильоту.
    5. Поясніть, чи достатньо простої імпутації для такої ознаки, як `DepTime`.

- **Підказки:** перед порівнянням зафіксуйте чисту контрольну модель без штучних пропусків, щоб мати референтний рівень якості.

---

**Варіант 2 – Екстремальні викиди в `Distance`:**

- **Мета:** оцінити, як вилучення або winsorization впливають на модель за наявності нетипових відстаней.

- **Кроки:**
    1. Дослідіть розподіл `Distance` і визначте порогові значення для дуже малих та дуже великих відстаней.
    2. Сформуйте три сценарії: без обробки, з видаленням викидів і з clipping або winsorization на обраних перцентилях.
    3. Навчіть однакові моделі на кожному сценарії й зафіксуйте метрики та кількість збережених рядків.
    4. Порівняйте якість, а також зміну середнього і стандартного відхилення `Distance` після кожного способу обробки.
    5. Зробіть висновок, чи є видалення викидів виправданим, чи краще використовувати м’якше обмеження значень.

- **Підказки:** коментуйте не лише приріст або падіння метрики, а й те, чи не вилучили ви рідкісні, але реальні рейси.

---

**Варіант 3 – Некоректні значення `Origin` і `Dest`:**

- **Мета:** перевірити, як обробка пошкоджених кодів аеропортів впливає на модель.

- **Кроки:**
    1. Перевірте `Origin` і `Dest` на порожні рядки, коди неправильної довжини або явно нетипові значення.
    2. Реалізуйте дві стратегії: заміну на спеціальну категорію `UNK` та вилучення проблемних рядків.
    3. Для кожної стратегії підготуйте однаковий набір ознак і спосіб кодування категорій.
    4. Навчіть моделі та порівняйте метрики, кількість збережених записів і стабільність важливостей маршрутних ознак.
    5. Поясніть, чи доцільніше зберігати проблемні записи через `UNK`, ніж видаляти їх повністю.

- **Підказки:** якщо вводите `UNK`, переконайтеся, що таке значення коректно з’являється і в train, і в test після застосування того самого правила.

---

**Варіант 4 – Локальна інтерполяція `DepTime` у межах `Origin`:**

- **Мета:** перевірити, чи допомагає локальне відновлення часу вильоту краще за глобальну медіану.

- **Кроки:**
    1. Відсортуйте рейси за `Origin`, `Month`, `DayofMonth` і `DepTime`, щоб отримати квазіхронологічний порядок усередині аеропорту.
    2. Створіть контрольований набір із пропусками в `DepTime`, а далі реалізуйте локальну інтерполяцію на основі сусідніх валідних рейсів у тому самому `Origin`.
    3. Побудуйте контрольну модель з глобальною медіанною імпутацією та альтернативну модель з локальною інтерполяцією.
    4. Порівняйте якість обох підходів і окремо наведіть приклади, де локальна інтерполяція дала явно реалістичніші значення.
    5. Поясніть, чи виправдана складність локальної імпутації у цій задачі.

- **Підказки:** якщо для певного `Origin` бракує сусідів, заздалегідь визначте резервне правило, наприклад використання глобальної медіани.

---

**Варіант 5 – Аномальні значення `DayofMonth`:**

- **Мета:** дослідити, як модель реагує на некоректні номери дня місяця.

- **Кроки:**
    1. Виявте всі записи, де `DayofMonth` виходить за межі допустимого діапазону, або навмисно зімітуйте невелику частку таких значень для експерименту.
    2. Реалізуйте стратегію вилучення і стратегію виправлення, наприклад заміну на найближче допустиме значення.
    3. Для кожної стратегії навчіть модель за однакових умов та зафіксуйте метрики.
    4. Порівняйте, як змінилася якість і чи не спотворився календарний розподіл спостережень після корекції.
    5. Зробіть висновок, чи краще виправляти такі помилки, чи не ризикуєте ви вносити штучний шум.

- **Підказки:** пояснюйте правило корекції формально; довільна заміна без чіткої логіки не вважається якісним очищенням.

---

**Варіант 6 – Ітеративна імпутація числових ознак:**

- **Мета:** порівняти складнішу багатокрокову імпутацію з простим заповненням медіаною.

- **Кроки:**
    1. Сформуйте контрольований набір з пропусками у `DepTime` і `Distance`, аби забезпечити відтворюваний експеримент.
    2. Реалізуйте базову медіанну імпутацію та окремо `IterativeImputer` або іншу ітеративну схему для числових полів.
    3. Побудуйте для кожної версії набору однакову `XGBoost`-модель та оцініть її на спільній validation-вибірці.
    4. Порівняйте метрики, час підготовки даних і правдоподібність відновлених значень.
    5. Зробіть висновок, чи виправдана складніша імпутація для цього набору даних.

- **Підказки:** якщо використовуєте ітеративну імпутацію, застосовуйте її лише на train-частині й переносіть fitted-імпутер на validation/test.

---

**Варіант 7 – Індикатори пропущених значень як окремі ознаки:**

- **Мета:** перевірити, чи є сам факт відсутності значення прогностично значущим.

- **Кроки:**
    1. Визначте поля, у яких виникають або штучно створюються пропуски, наприклад `DepTime` і `Distance`.
    2. Для кожного такого поля створіть бінарний прапорець `isNA`, а числове значення заповніть константою або медіаною.
    3. Навчіть контрольну модель без індикаторів та модель з індикаторами пропусків.
    4. Порівняйте метрики і перевірте, чи входять прапорці `isNA` до числа важливих ознак.
    5. Поясніть, чи сигналізує відсутність даних про підвищений операційний ризик або лише про технічну помилку.

- **Підказки:** індикатор пропуску має створюватися до імпутації, інакше сам факт пропуску буде втрачено.

---

**Варіант 8 – Відсутня `Distance` для окремих перевізників:**

- **Мета:** дослідити поведінку моделі в сценарії систематично неповних даних для підгрупи рейсів.

- **Кроки:**
    1. Зімітуйте ситуацію, коли для кількох значень `UniqueCarrier` відсутня `Distance`, або знайдіть природно проблемну підгрупу в даних.
    2. Порівняйте три підходи: повне вилучення цих рейсів, імпутацію `Distance` та окрему резервну модель без `Distance`.
    3. Для резервного сценарію сформуйте правило маршрутизації: коли `Distance` невідома, застосовуйте спеціалізовану модель.
    4. Оцініть якість кожного підходу на спільній validation-вибірці та покажіть, як змінюється покриття даних.
    5. Зробіть висновок, чи варто будувати окремий сценарій обробки для систематично неповних даних.

- **Підказки:** у звіті окремо вкажіть, як змінився розподіл `UniqueCarrier` після кожної стратегії.

---

**Варіант 9 – Winsorization для аномального `DepTime`:**

- **Мета:** перевірити, чи допомагає жорстке обмеження значень часу вильоту боротися з викидами.

- **Кроки:**
    1. Виявте неможливі значення `DepTime`, наприклад від’ємні коди, значення понад `2359` або некоректні хвилини.
    2. Створіть два набори: з видаленням проблемних рядків і з winsorization або clipping часу до допустимих меж.
    3. Навчіть однакові моделі на обох варіантах і порівняйте їх з контрольним набором без спеціальної обробки.
    4. Проаналізуйте, як змінилася частка затримок у записах з аномальним `DepTime` та в усьому наборі загалом.
    5. Зробіть висновок, чи корекція часу краще зберігає інформацію, ніж повне видалення рядків.

- **Підказки:** якщо застосовуєте clipping, окремо поясніть, чому це краще за довільну заміну усіх аномалій однією константою.

---

**Варіант 10 – Категорія `UNK` для проблемного `DayOfWeek`:**

- **Мета:** оцінити, чи є сенс виділяти невизначені значення дня тижня в окрему категорію.

- **Кроки:**
    1. Знайдіть або змоделюйте записи з некоректним `DayOfWeek`, які не можна надійно перетворити на валідне числове значення.
    2. Створіть дві схеми обробки: заміну на `UNK` і повне вилучення проблемних рядків.
    3. Закодуйте обидва варіанти та навчіть для них порівнювані моделі.
    4. Перевірте, чи є `UNK` серед важливих ознак і як цей підхід впливає на метрики.
    5. Поясніть, чи слід трактувати невизначений день тижня як окремий сигнал проблеми в даних.

- **Підказки:** не змішуйте `UNK` з валідним числовим кодом дня; це має бути окрема категорія.

---

**Варіант 11 – Перевірка календарної узгодженості `Month` і `DayofMonth`:**

- **Мета:** виявити і виправити суперечливі календарні пари в даних.

- **Кроки:**
    1. Після очищення `Month` і `DayofMonth` перевірте, чи не зустрічаються неможливі комбінації на кшталт 31-го дня для короткого місяця.
    2. Реалізуйте два сценарії: вилучення некоректних записів і корекцію до найближчого допустимого дня.
    3. Для кожного сценарію побудуйте однаковий набір ознак і навчіть `XGBoost`.
    4. Порівняйте метрики, кількість виправлених рядків і вплив на розподіл спостережень за місяцями.
    5. Зробіть висновок, чи виправдання таких помилок покращує дані, чи вносить додатковий шум.

- **Підказки:** якщо обираєте корекцію, формалізуйте правило окремо для місяців з 30 і 31 днем.

---

**Варіант 12 – Стійкість моделі до шуму в мітках класу:**

- **Мета:** перевірити, наскільки `XGBoost` вразлива до помилок у `dep_delayed_15min`.

- **Кроки:**
    1. Візьміть очищений train і штучно змініть невелику частку міток `Y` на `N` або навпаки, зафіксувавши частку шуму та random seed.
    2. Навчіть контрольну модель на чистих даних і модель на зашумленій версії train.
    3. Оцініть обидві моделі на незмінній validation-вибірці без шуму в цілі.
    4. Порівняйте метрики та проаналізуйте, які типи помилок зросли найбільше після внесення шуму.
    5. Зробіть висновок, наскільки критичними є помилки в цільовій змінній для цієї задачі.

- **Підказки:** не шуміть validation-частину, інакше ви не зможете коректно оцінити деградацію моделі.

---

**Варіант 13 – «М’які» мітки для неоднозначних спостережень:**

- **Мета:** перевірити експериментальний підхід до навчання на невпевнених прикладах.

- **Кроки:**
    1. Визначте групу потенційно неоднозначних записів, наприклад із проблемним `DepTime`, підозрілими календарними комбінаціями або пограничними випадками після очищення.
    2. Призначте цим записам проміжні значення цілі, наприклад `0.2`, `0.5` або `0.8`, і підготуйте контрольний варіант з жорсткою бінарною ціллю.
    3. Навчіть експериментальну модель у режимі регресії або probability regression і перетворіть прогноз назад на бінарне рішення порогом.
    4. Порівняйте результат з класичним бінарним `XGBoost` на спільній validation-вибірці.
    5. Поясніть, чи є «м’які» мітки корисним способом поводження з невпевненими спостереженнями у цій лабораторній роботі.

- **Підказки:** у звіті чітко відокремте цей експериментальний підхід від стандартної постановки задачі бінарної класифікації.

---

**Варіант 14 – Подвійні пропуски `DepTime` і `Distance` з позначкою невизначеності:**

- **Мета:** перевірити, чи допомагає індикатор високої невизначеності при складній імпутації.

- **Кроки:**
    1. Створіть або знайдіть підмножину рядків, де одночасно відсутні `DepTime` і `Distance`, та сформуйте для неї окремий список індексів.
    2. Реалізуйте складнішу імпутацію для цих двох полів, наприклад ітеративний підхід або ієрархічне заповнення за `Origin` і `UniqueCarrier`.
    3. Додайте бінарний прапорець `isHighUncertainty`, який позначатиме записи з подвійним відновленням.
    4. Порівняйте модель з цією ознакою та модель без неї, зберігаючи однакову схему оцінювання.
    5. Зробіть висновок, чи несе маркер високої невизначеності самостійну прогностичну інформацію.

- **Підказки:** якщо імпутація для частини рядків виглядає надто штучною, окремо проаналізуйте сценарій їх вилучення як контрольну альтернативу.

---

**Варіант 15 – `RobustScaler` проти інших способів масштабування:**

- **Мета:** перевірити, чи має робастне масштабування практичний ефект для ознак із потенційними викидами.

- **Кроки:**
    1. Виберіть числові ознаки, насамперед `DepTime` і `Distance`, і дослідіть їхній розподіл на наявність викидів.
    2. Підготуйте кілька версій даних: без масштабування, зі `StandardScaler`, з `MinMaxScaler` і з `RobustScaler`.
    3. Навчіть однакову `XGBoost`-модель на кожній версії й зафіксуйте ключові метрики.
    4. Порівняйте, чи змінюється якість і наскільки стабільними стають результати після робастного масштабування.
    5. Поясніть, чи варто включати масштабування до пайплайна для tree-based моделі саме у цьому наборі даних.

- **Підказки:** основна цінність варіанта полягає не в механічному порівнянні чисел, а в аргументованому поясненні результату.

---

**Варіант 16 – Часткове пошкодження календарних полів:**

- **Мета:** оцінити стійкість моделі до шуму в `Month` і `DayOfWeek`.

- **Кроки:**
    1. На копії train випадково замініть невелику частку значень `Month` або `DayOfWeek` на інші валідні коди, зафіксувавши частку корупції.
    2. Додайте бінарну ознаку `isCorrupted`, яка позначатиме штучно змінені рядки.
    3. Навчіть модель на зашумлених даних без прапорця та окремо модель з прапорцем `isCorrupted`.
    4. Оцініть обидві моделі на чистій validation-вибірці та порівняйте втрату якості від корупції.
    5. Поясніть, чи може модель частково компенсувати пошкодження, якщо вона отримує явний сигнал про недостовірність запису.

- **Підказки:** validation має залишатися незмінною; інакше ви не відокремите ефект шуму в train від шуму в оцінюванні.

---

**Варіант 17 – IQR-відсікання для числових викидів:**

- **Мета:** перевірити класичне правило `1.5 * IQR` для очищення `DepTime` і `Distance`.

- **Кроки:**
    1. Обчисліть `Q1`, `Q3` і `IQR` для `DepTime` та `Distance`, а потім визначте межі допустимих значень за правилом `1.5 * IQR`.
    2. Побудуйте очищений набір, у якому рядки поза межами вилучено, і збережіть статистику про кількість вилучених записів.
    3. Навчіть контрольну модель на повному наборі й альтернативну модель на очищеному наборі.
    4. Порівняйте якість, а також окремо проаналізуйте, які типи записів найчастіше вилучалися за правилом IQR.
    5. Зробіть висновок, чи не є правило IQR надто агресивним для цієї прикладної задачі.

- **Підказки:** для `DepTime` оцініть, чи не конфліктує статистичне правило IQR з предметно-обумовленими межами часу від `0` до `2359`.

---

**Варіант 18 – Поєднання очищення від викидів з tuning:**

- **Мета:** перевірити, чи змінюються оптимальні гіперпараметри після усунення викидів.

- **Кроки:**
    1. Побудуйте контрольний набір без спеціальної фільтрації та альтернативний набір після обробки викидів `Distance` або `DepTime`.
    2. На кожному наборі виконайте однаковий tuning, наприклад за `max_depth` і `n_estimators`, використавши ту саму метрику та CV-схему.
    3. Збережіть найкращі параметри, validation-результат і кількість рядків у кожному сценарії.
    4. Порівняйте, чи змінилися оптимальні параметри складності після очищення від викидів.
    5. Зробіть висновок, чи допомагає очищення моделі знайти кращу конфігурацію, чи tuning і без того компенсує шум.

- **Підказки:** важливо аналізувати не лише найкращу метрику, а й зміну самої структури знайденої моделі.

---

**Варіант 19 – Модульне виправлення `DepTime` за правилом `% 2400`:**

- **Мета:** перевірити, чи може формальна корекція некоректного часу бути кращою за видалення рядків.

- **Кроки:**
    1. Визначте підмножину записів з `DepTime > 2400` або змоделюйте такий тип помилки у контрольованій частці рядків.
    2. Реалізуйте два сценарії: повне вилучення проблемних записів і корекцію через операцію `DepTime % 2400` з подальшою перевіркою хвилин.
    3. Для кожного сценарію побудуйте однакову модель і оцініть її на незмінній validation-вибірці.
    4. Порівняйте метрики та наведіть приклади, коли модульна корекція дає правдоподібний або, навпаки, сумнівний результат.
    5. Поясніть, чи може таке формальне виправлення бути частиною робочого пайплайна очищення.

- **Підказки:** після операції `% 2400` обов’язково перевірте, чи хвилинна частина нового часу є допустимою.

---

**Варіант 20 – Єдина функція `clean_flight_data(df)` для пропусків і викидів:**

- **Мета:** оформити результати очищення у відтворюваний і придатний до повторного використання пайплайн.

- **Кроки:**
    1. Узагальніть усі обрані вами правила обробки пропусків, аномальних кодів, викидів і категоріальних проблем у функцію `clean_flight_data(df)`.
    2. Застосуйте функцію окремо до train і test, зафіксувавши статистику очищення для кожного набору.
    3. Побудуйте модель на очищеному train і оцініть її на validation, використавши ті самі ознаки і параметри, що в контрольному сценарії.
    4. Порівняйте результат із моделлю без єдиного пайплайна очищення та покажіть, які саме елементи функції дали найбільший ефект.
    5. Поясніть, чому оформлення правил у єдину функцію є принципово важливим для промислового сценарію використання моделі.

- **Підказки:** функція має бути детермінованою, прозорою і не повинна використовувати цільову змінну для очищення test set.

---

<a class="anchor" id="lab-10.5"></a>

## <span style="color:blue; font-size:1.5em;">10.5. Розширені ансамблеві стратегії та інтерпретація</span>

[Назад до змісту](#lab-10)


### <span style="color:red; font-size:1.5em;">Завдання 5</span>

**Для всіх варіантів:**

- Побудуйте сильну референтну модель `XGBoost` на очищених даних і порівнюйте кожен складніший підхід саме з нею, а не з довільною слабкою baseline-конфігурацією.
- Якщо використовуєте `LightGBM`, `CatBoost`, `SHAP`, `LIME`, калібрування або ансамблювання, чітко описуйте вхідні дані кожного етапу, спосіб кодування ознак і механізм контролю leakage.
- Для ансамблів і постобробки подавайте як мінімум метрики якості, правило обчислення ваг або порогів та короткий висновок про практичну придатність підходу.
- В інтерпретаційних і fairness-варіантах робіть висновки лише на підставі спостережених результатів для `Origin`, `Dest`, `UniqueCarrier`, `DepTime`, `Distance` та інших наявних ознак, не підмінюючи аналіз припущеннями.

---

**Варіант 1 – Блендинг `LightGBM` і `XGBoost`:**

- **Мета:** перевірити, чи дає комбінація двох бустингових реалізацій кращий результат за кожну окремо.

- **Кроки:**
    1. Підготуйте однаковий очищений набір ознак для `LightGBM` і `XGBoost`, узгодивши спосіб кодування категоріальних полів.
    2. Навчіть обидві моделі окремо і отримайте їхні ймовірності на спільній validation-вибірці.
    3. Побудуйте бленд через просте або зважене середнє ймовірностей, зафіксувавши формулу ваг.
    4. Порівняйте `roc_auc_score`, `f1_score` та інші релевантні метрики для кожної окремої моделі й для ансамблю.
    5. Зробіть висновок, чи виправданий блендинг двох бустингових алгоритмів у цій задачі.

- **Підказки:** не оцінюйте якість на `flight_delays_test.csv`, оскільки там немає `dep_delayed_15min`; використовуйте validation або out-of-fold прогнози.

---

**Варіант 2 – Дворівневий stacking із метамоделлю:**

- **Мета:** дослідити, чи вміє метамодель поєднати сильні й слабші базові алгоритми краще за просте усереднення.

- **Кроки:**
    1. Виберіть кілька базових моделей, наприклад `XGBoost`, `RandomForest` і `LogisticRegression`, та підготуйте для них спільний набір ознак.
    2. Отримайте out-of-fold прогнози кожної базової моделі на train через крос-валідацію.
    3. Використайте ці прогнози як ознаки для метамоделі, наприклад другого рівня `XGBoost` або `LogisticRegression`.
    4. Оцініть стекований ансамбль на validation і порівняйте його з найкращою окремою базовою моделлю.
    5. Поясніть, чи виправдана додаткова складність stacking-підходу для задачі прогнозування затримок.

- **Підказки:** базові моделі мають генерувати прогнози для train лише out-of-fold способом, інакше метамодель отримає leakage.

---

**Варіант 3 – Cost-sensitive навчання для виявлення затримок:**

- **Мета:** зменшити кількість пропущених затриманих рейсів за рахунок асиметричного штрафу.

- **Кроки:**
    1. Зафіксуйте референтну `XGBoost`-модель і визначте, який тип помилки в цій задачі є дорожчим – false negative чи false positive.
    2. Реалізуйте cost-sensitive підхід через `scale_pos_weight`, custom objective або інший доступний механізм асиметричного штрафу.
    3. Навчіть стандартну і cost-sensitive-модель на одному й тому самому наборі ознак та split.
    4. Порівняйте `recall`, `precision`, `f1_score` і зміну confusion matrix для класу затриманих рейсів.
    5. Поясніть, чи виправдане підсилення ваги позитивного класу з погляду практичного застосування моделі.

- **Підказки:** якщо не реалізуєте custom loss, прямо вкажіть, що `scale_pos_weight` є практичним наближенням до cost-sensitive навчання.

---

**Варіант 4 – `CatBoost` проти закодованого `XGBoost`:**

- **Мета:** порівняти ручне кодування категорій із вбудованою підтримкою категоріальних ознак.

- **Кроки:**
    1. Підготуйте два сценарії: `CatBoost` з передачею `UniqueCarrier`, `Origin` і `Dest` як категоріальних ознак та `XGBoost` з ручним кодуванням цих полів.
    2. Узгодьте train/validation split і максимальну складність моделей, щоб порівняння було чесним.
    3. Навчіть обидві моделі і зафіксуйте якість, час навчання та простоту підготовки даних.
    4. Порівняйте, чи дає `CatBoost` перевагу у якості або в організації пайплайна.
    5. Зробіть висновок, який інструмент виглядає переконливішим для задачі з великою кількістю категорій.

- **Підказки:** якщо використовуєте `CatBoost`, не дублюйте ручне OHE для тих самих полів; інакше зникне сенс порівняння.

---

**Варіант 5 – Bagging над `XGBoost` через bootstrap-вибірки:**

- **Мета:** перевірити, чи зменшує усереднення кількох бустингових моделей дисперсію прогнозу.

- **Кроки:**
    1. Згенеруйте кілька bootstrap-вибірок із train однакового розміру та зафіксуйте кількість моделей в ансамблі.
    2. Навчіть окрему `XGBoost`-модель на кожній bootstrap-вибірці з однаковими гіперпараметрами.
    3. Усередніть імовірності на validation і сформуйте підсумковий ансамблевий прогноз.
    4. Порівняйте ансамбль із однією моделлю, натренованою на всьому train, за метриками якості та стабільністю.
    5. Поясніть, чи додає bootstrap-ансамбль практичну цінність поверх уже сильного бустингового алгоритму.

- **Підказки:** у звіті наведіть не лише метрики, а й короткий коментар про обчислювальну ціну такого ансамблю.

---

**Варіант 6 – Допоміжна модель календарного ризику як мета-ознака:**

- **Мета:** зберегти ідею багатоступеневого моделювання, не переходячи до штучної multi-output постановки.

- **Кроки:**
    1. Побудуйте допоміжну модель, яка використовує лише календарні ознаки `Month`, `DayofMonth`, `DayOfWeek` і повертає імовірність затримки `calendarRisk`.
    2. Отримайте out-of-fold значення `calendarRisk` для train та коректні прогнози для validation/test.
    3. Додайте `calendarRisk` як нову мета-ознаку до основної моделі `XGBoost`, що вже використовує повний набір ознак.
    4. Порівняйте основну модель з і без цієї мета-ознаки за однакової схеми оцінювання.
    5. Поясніть, чи допомагає окреме моделювання календарного ризику краще структурувати інформацію для основної моделі.

- **Підказки:** `calendarRisk` для train має будуватися out-of-fold способом; інакше основна модель отримає штучно завищений сигнал.

---

**Варіант 7 – Місячно-спеціалізовані моделі:**

- **Мета:** перевірити, чи виграє модель від сезонної спеціалізації за місяцями.

- **Кроки:**
    1. Після очищення `Month` розбийте train на піднабори за місяцями та підготуйте окремі моделі для кожної достатньо великої групи.
    2. Паралельно навчіть одну глобальну модель на всіх даних як референтний варіант.
    3. На validation застосуйте правило маршрутизації за місяцем: запис оцінює або спеціалізована, або глобальна модель.
    4. Порівняйте якість місячно-спеціалізованого підходу з глобальною моделлю та проаналізуйте, у яких місяцях спеціалізація найбільше допомагає.
    5. Зробіть висновок, чи виправдана підтримка окремих моделей за сезонами.

- **Підказки:** якщо для окремих місяців даних замало, передбачте повернення до глобальної моделі і зафіксуйте це правило.

---

**Варіант 8 – Монотонні обмеження в `XGBoost` або `LightGBM`:**

- **Мета:** оцінити, чи допомагають структурні обмеження зробити модель стабільнішою та інтерпретованішою.

- **Кроки:**
    1. Виберіть одну або дві числові ознаки, для яких можна обґрунтувати очікувану монотонну тенденцію або, навпаки, заборону на необмежене зростання ризику.
    2. Налаштуйте модель з `monotone_constraints` або відповідним механізмом у вибраному бустинговому фреймворку.
    3. Навчіть обмежену і необмежену версії моделі на однаковому наборі ознак.
    4. Порівняйте якість, а також форму залежності прогнозу від ознаки через `PDP` або аналогічну візуалізацію.
    5. Поясніть, чи покращили обмеження інтерпретованість без значної втрати якості.

- **Підказки:** вводьте обмеження лише для тих ознак, де маєте предметно обґрунтовану гіпотезу, а не механічно для всіх числових полів.

---

**Варіант 9 – Текстове представлення маршруту через `TF-IDF`:**

- **Мета:** адаптувати ідею текстових ознак до наявних колонок без залучення зовнішніх даних.

- **Кроки:**
    1. Створіть текстовий рядок на основі `Origin`, `Dest` і `UniqueCarrier`, наприклад `Origin_Dest_UniqueCarrier`, для кожного запису.
    2. Перетворіть ці рядки на векторне представлення через `TfidfVectorizer` або інший компактний текстовий метод.
    3. Об’єднайте TF-IDF-ознаки з числовим блоком і навчіть модель на розширеному просторі ознак.
    4. Порівняйте результат з моделлю, де ті самі поля подано стандартним категоріальним кодуванням.
    5. Зробіть висновок, чи дає текстове представлення маршруту додатковий виграш у цій задачі.

- **Підказки:** цей варіант не потребує зовнішніх текстів; достатньо осмислено перетворити наявні кодові поля на символьні комбінації.

---

**Варіант 10 – Глобальна і локальна інтерпретація через `SHAP`:**

- **Мета:** пояснити поведінку сильної моделі як на рівні всього набору, так і на рівні окремих рейсів.

- **Кроки:**
    1. Навчіть сильну `XGBoost`-модель на очищеному наборі ознак і збережіть validation-прогнози.
    2. Обчисліть `SHAP`-значення для validation-частини та побудуйте підсумковий графік `SHAP`.
    3. Виберіть кілька рейсів із високою та низькою прогнозною імовірністю затримки і побудуйте для них локальні пояснення.
    4. Порівняйте глобальні й локальні висновки, звернувши увагу на роль `DepTime`, `Distance`, `UniqueCarrier`, `Origin` і `Dest`.
    5. Поясніть, як `SHAP` допомагає перевірити здоровий глузд моделі та знайти нетипові рішення.

- **Підказки:** добре, якщо серед локально проаналізованих прикладів буде хоча б один false positive або false negative.

---

**Варіант 11 – Порівняння `LIME` і `SHAP` для локальних пояснень:**

- **Мета:** зіставити дві популярні техніки локальної інтерпретації однієї й тієї самої моделі.

- **Кроки:**
    1. Навчіть одну сильну бустингову модель і виберіть кілька показових прикладів з validation: затримані, незатримані та помилково класифіковані.
    2. Згенеруйте для тих самих прикладів локальні пояснення через `LIME` і `SHAP`.
    3. Порівняйте, які ознаки обидва методи вважають найважливішими та де їхні висновки розходяться.
    4. Оцініть, який із методів дає більш стабільне та зрозуміле пояснення для ваших прикладів.
    5. Зробіть висновок, який інструмент ви б використали для пояснення рішень моделі викладачеві або замовнику.

- **Підказки:** обмежтеся кількома добре підібраними прикладами; масове порівняння локальних пояснень швидко стає шумним і важким для інтерпретації.

---

**Варіант 12 – Fairness-аналіз за `Origin` або `UniqueCarrier`:**

- **Мета:** перевірити, чи не робить модель систематично нерівних помилок для окремих груп рейсів.

- **Кроки:**
    1. Навчіть сильну референтну модель і отримайте прогнози на validation-вибірці.
    2. Згрупуйте записи за `Origin`, `UniqueCarrier` або іншою змістовною категорією з достатнім розміром групи.
    3. Для кожної групи обчисліть принаймні `false positive rate`, `false negative rate` і частку позитивних прогнозів.
    4. Порівняйте групи між собою, знайдіть найбільші розбіжності та перевірте, чи не пов’язані вони з дисбалансом обсягу даних.
    5. Поясніть, як ці результати слід інтерпретувати у навчальному контексті без поспішних нормативних висновків.

- **Підказки:** fairness-аналіз має базуватися на достатньо великих групах; дрібні категорії краще об’єднувати або виключати з підсумкового порівняння.

---

**Варіант 13 – Калібрування ймовірностей через `Platt scaling` або isotonic regression:**

- **Мета:** оцінити, чи робить калібрування прогнозні ймовірності більш надійними.

- **Кроки:**
    1. Навчіть сильну `XGBoost`-модель і отримайте її неочищені ймовірності на validation-вибірці.
    2. Використайте `CalibratedClassifierCV` або інший коректний механізм калібрування через `sigmoid` чи `isotonic`.
    3. Побудуйте calibration curve або обчисліть `Brier score` для неочищеної і каліброваної моделей.
    4. Порівняйте не лише узагальнені метрики, а й те, як змінився розподіл прогнозних ймовірностей.
    5. Зробіть висновок, чи варто калібрувати модель перед використанням порогових правил або ризикових інтерпретацій.

- **Підказки:** калібрування треба виконувати на відкладених даних або через крос-валідаційний механізм, а не на тих самих даних, де модель щойно навчалася.

---

**Варіант 14 – Аналіз найнепевніших прогнозів:**

- **Мета:** виявити системні причини невпевненості моделі та спробувати їх зменшити.

- **Кроки:**
    1. Навчіть сильну модель і знайдіть на validation записи з прогнозами поблизу порога, наприклад у діапазоні `0.45–0.55`.
    2. Проаналізуйте розподіл їхніх ознак, зокрема `DepTime`, `Distance`, `Origin`, `Dest` і `UniqueCarrier`, та порівняйте з упевненими прикладами.
    3. Виділіть один або два повторювані патерни невпевненості й запропонуйте для них нові похідні ознаки або перетворення.
    4. Перенавчіть модель із цими доповненнями та оцініть, чи зменшилася кількість невпевнених випадків або покращилася метрика.
    5. Поясніть, що саме говорить зона невпевненості про обмеження поточного набору ознак.

- **Підказки:** не обмежуйтеся якісним описом; покажіть чисельно, яка частка validation-випадків потрапляє до вашої «сірої зони».

---

**Варіант 15 – Три рівні ризику замість жорсткого бінарного рішення:**

- **Мета:** перетворити бінарний прогноз на більш практичну систему рішень `low / medium / high risk`.

- **Кроки:**
    1. Навчіть та за потреби калібруйте сильну `XGBoost`-модель, щоб працювати з надійнішими ймовірностями затримки.
    2. Визначте два пороги, які розбивають прогноз на три зони ризику: низьку, проміжну та високу.
    3. Для кожної зони обчисліть фактичну частку затримок, чисельність записів та середню імовірність прогнозу.
    4. Порівняйте цей триступеневий режим із класичним бінарним правилом `threshold = 0.5` з точки зору інтерпретованості та практичної дії.
    5. Поясніть, коли багаторівнева шкала ризику може бути кориснішою за жорстку двійкову класифікацію.

- **Підказки:** добирайте пороги не довільно, а на підставі calibration curve, компромісу між precision та recall або логіки операційного використання.

---

**Варіант 16 – Зсув порога класифікації:**

- **Мета:** знайти поріг, який краще відповідає вибраній бізнес-меті моделі.

- **Кроки:**
    1. Навчіть сильну модель і отримайте ймовірності затримки на validation.
    2. Переберіть низку порогів класифікації, наприклад від `0.30` до `0.70` з фіксованим кроком.
    3. Для кожного порога обчисліть `precision`, `recall`, `f1_score` та за потреби частку позитивних прогнозів.
    4. Побудуйте таблицю або графік `precision – recall` і визначте поріг, що найкраще відповідає вашому критерію.
    5. Поясніть, чи є стандартний поріг `0.5` оптимальним для задачі прогнозування затримок.

- **Підказки:** якщо обираєте поріг за `f1_score`, окремо покажіть, що зміниться, якщо пріоритетом стане високий `recall`.

---

**Варіант 17 – Покрокове донавчання через часові батчі:**

- **Мета:** перевірити практичну ідею staged retraining замість недоступного повноцінного incremental learning.

- **Кроки:**
    1. Розбийте train у часовому порядку на дві або три послідовні частини за `Month`, `DayofMonth` і `DepTime`.
    2. Навчіть першу модель на ранньому батчі, а потім окремо перенавчіть або розширте сценарій, додаючи наступні часові батчі.
    3. Для кожної стадії оцініть модель на спільній пізнішій validation-частині.
    4. Порівняйте staged retraining з моделлю, що одразу навчена на всьому train.
    5. Зробіть висновок, чи має сенс періодичне оновлення моделі в умовах надходження нових історичних даних.

- **Підказки:** у цьому варіанті головне порівняти стратегію оновлення моделі, а не стверджувати наявність справжнього online-learning у `XGBoost`.

---

**Варіант 18 – Журналювання помилково класифікованих рейсів:**

- **Мета:** перетворити помилки моделі на джерело системного поліпшення ознак і пайплайна.

- **Кроки:**
    1. Навчіть сильну модель і для validation збережіть списки false positive та false negative разом з основними ознаками записів.
    2. Побудуйте короткий лог помилок у табличній формі та виділіть повторювані патерни за `Origin`, `Dest`, `UniqueCarrier`, `DepTime` і `Distance`.
    3. Визначте одну або дві гіпотези, які пояснюють повторні помилки, наприклад нестачу маршрутної ознаки або невдалий поріг.
    4. Модифікуйте модель або набір ознак відповідно до гіпотез і перевірте, чи зменшилася кількість типових помилок.
    5. Поясніть, чому системне журналювання помилок є важливішим за разовий перегляд окремих неправильних прогнозів.

- **Підказки:** лог має бути достатньо структурованим, щоб за ним можна було швидко знайти повторювані проблеми в моделі.

---

**Варіант 19 – Динамічний поріг за годиною вильоту:**

- **Мета:** перевірити, чи корисно робити більш гнучке рішення для різних часових інтервалів.

- **Кроки:**
    1. Навчіть сильну модель і отримайте ймовірності затримки на validation, додавши до аналізу `DepHour`.
    2. Розбийте рейси на кілька часових зон, наприклад пікові ранкові години, денний інтервал і вечірній блок.
    3. Для кожної зони підберіть власний поріг класифікації за однаковим критерієм, наприклад максимізацією `f1_score` або контролем `recall`.
    4. Порівняйте динамічний підхід з єдиним глобальним порогом на всій validation-вибірці.
    5. Зробіть висновок, чи виправдовує локальний tuning порогів додаткову складність правила прийняття рішення.

- **Підказки:** пороги слід добирати лише на validation або в окремому внутрішньому циклі, а не на test set без міток.

---

**Варіант 20 – Підсумковий production-ready pipeline:**

- **Мета:** звести очищення, feature engineering, tuning, калібрування й інтерпретацію в один завершений сценарій роботи.

- **Кроки:**
    1. Опишіть і реалізуйте повний порядок дій: завантаження `train` і `test`, очищення, кодування, створення ознак, поділ на train/validation, tuning моделі.
    2. Додайте до пайплайна щонайменше один елемент advanced-рівня: ансамблювання, калібрування, `SHAP`-аналіз або структуроване журналювання помилок.
    3. Навчіть фінальну модель, оцініть її на validation і сформуйте короткий звіт із ключовими метриками та сильними/слабкими сторонами.
    4. Покажіть, як той самий пайплайн без зміни логіки застосовується до `flight_delays_test.csv` для отримання прогнозів.
    5. Поясніть, які етапи цього пайплайна є критичними для стабільного використання моделі поза межами лабораторної роботи.

- **Підказки:** production-ready у межах цієї лабораторної роботи означає відтворюваність, узгодженість трансформацій та зрозумілість усіх рішень, а не розгортання сервісу.
